<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Add count of genotypes to INFO field — addCountGenos-methods • TVTB</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Add count of genotypes to INFO field — addCountGenos-methods"><meta property="og:description" content='Adds the total occurences of a set of genotypes as an INFO
field for each variant.
All given genotypes are counted toward a single
total (e.g. grand total of c("0/0", "0|0")),
while other genotypes are silently ignored.'><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Add count of genotypes to INFO field</h1>

    <div class="hidden name"><code>addCountGenos-methods.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Adds the total occurences of a set of genotypes as an INFO
field for each variant.
All given genotypes are counted toward a single
total (<em>e.g.</em> grand total of <code>c("0/0", "0|0")</code>),
while other genotypes are silently ignored.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for class 'ExpandedVCF'</span></span>
<span><span class="fu">addCountGenos</span><span class="op">(</span></span>
<span>        <span class="va">vcf</span>, <span class="va">genos</span>, <span class="va">key</span>, <span class="va">description</span>,</span>
<span>        samples <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span>, force <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <p></p>
<dl><dt id="arg-vcf">vcf<a class="anchor" aria-label="anchor" href="#arg-vcf"></a></dt>
<dd><p><code>ExpandedVCF</code> object.</p></dd>

    <dt id="arg-genos">genos<a class="anchor" aria-label="anchor" href="#arg-genos"></a></dt>
<dd><p><code>character</code> vector of genotypes to count (toward a common unique total).</p></dd>

    <dt id="arg-key">key<a class="anchor" aria-label="anchor" href="#arg-key"></a></dt>
<dd><p>Name of the INFO field to create or update
(<code>character</code> vector of length 1). See <em>Details</em> below.</p></dd>

    <dt id="arg-description">description<a class="anchor" aria-label="anchor" href="#arg-description"></a></dt>
<dd><p><code>character</code> description of the INFO field to create or overwrite
(<code>character</code> vector of length 1).</p></dd>

    <dt id="arg-samples">samples<a class="anchor" aria-label="anchor" href="#arg-samples"></a></dt>
<dd><p><code>integer</code>, <code>numeric</code> or <code>character</code> vector
indicating samples to consider in <code>VariantAnnotation::geno(vcf)</code>.
If not specified, all samples are considered.</p></dd>

    <dt id="arg-force">force<a class="anchor" aria-label="anchor" href="#arg-force"></a></dt>
<dd><p>If <code>TRUE</code>, the field header and data will be overwritten if present;
If <code>FALSE</code>, an error is thrown if the field already exists.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><code>ExpandedVCF</code> object including an additional INFO field stating
    the count of genotypes.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Kevin Rue-Albrecht</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>In all cases, the new INFO field is inserted after the last existing field.
In other words, overwriting an existing INFO field is achieved by dropping
it from the data and header of the <code>info</code> slot,
and subsequently inserting the new data after the last remaining INFO field.</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="countGenos-methods.html">countGenos,ExpandedVCF-method</a></code>
and <code>geno,VCF-method</code></p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example data ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># VCF file</span></span></span>
<span class="r-in"><span><span class="va">vcfFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"moderate.vcf"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># TVTB parameters</span></span></span>
<span class="r-in"><span><span class="va">tparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="TVTBParam-class.html">TVTBparam</a></span><span class="op">(</span><span class="fu"><a href="Genotypes-class.html">Genotypes</a></span><span class="op">(</span>ref <span class="op">=</span> <span class="st">"0|0"</span>, het <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0|1"</span>, <span class="st">"1|0"</span><span class="op">)</span>, alt <span class="op">=</span> <span class="st">"1|1"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Pre-process variants</span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span><span class="va">vcfFile</span>, param <span class="op">=</span> <span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu">expand</span><span class="op">(</span><span class="va">vcf</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example usage ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">addCountGenos</span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">vcf</span>, <span class="fu"><a href="Genotypes-class.html">het</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span>,</span></span>
<span class="r-in"><span>    <span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">[</span><span class="st">"het"</span><span class="op">]</span>,</span></span>
<span class="r-in"><span>    <span class="st">"Number of heterozygous genotypes"</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer></div>






  </body></html>

