<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Genotypes and allele frequencies for a given phenotype level — addPhenoLevelFrequencies-methods • TVTB</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Genotypes and allele frequencies for a given phenotype level — addPhenoLevelFrequencies-methods"><meta property="og:description" content='

Adds genotypes counts
(reference homozygote, heterozygote, and alternate homozygote)
and allele frequencies (alternate and minor) calculated
in a group of samples associated with a given level of a given phenotype
as INFO fields in an ExpandedVCF object.
Multiple genotypes can be counted toward a single
frequency (e.g. combined c("0/0", "0|0") for
homozygote reference genotypes).'><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genotypes and allele frequencies for a given phenotype level</h1>

    <div class="hidden name"><code>addPhenoLevelFrequencies-methods.Rd</code></div>
    </div>

    <div class="ref-description">


<p>Adds genotypes counts
(reference homozygote, heterozygote, and alternate homozygote)
and allele frequencies (alternate and minor) calculated
in a group of samples associated with a given level of a given phenotype
as INFO fields in an <code>ExpandedVCF</code> object.
Multiple genotypes can be counted toward a single
frequency (<em>e.g.</em> combined <code>c("0/0", "0|0")</code> for
homozygote reference genotypes).</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="co"># S4 method for class 'ExpandedVCF'</span></span>
<span><span class="fu">addPhenoLevelFrequencies</span><span class="op">(</span></span>
<span>    <span class="va">vcf</span>, <span class="va">pheno</span>, <span class="va">level</span>, force <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <p></p>
<dl><dt id="arg-vcf">vcf<a class="anchor" aria-label="anchor" href="#arg-vcf"></a></dt>
<dd><p><code>ExpandedVCF</code> object.</p>
<p><code>metadata(vcf)[["TVTBparam"]]</code> must contain a
<code><a href="TVTBParam-class.html">TVTBparam</a></code> object.</p></dd>

    <dt id="arg-pheno">pheno<a class="anchor" aria-label="anchor" href="#arg-pheno"></a></dt>
<dd><p>Phenotype in <code>colnames(colData(vcf))</code>.</p></dd>

    <dt id="arg-level">level<a class="anchor" aria-label="anchor" href="#arg-level"></a></dt>
<dd><p>Phenotype level in <code>colData(vcf)[,pheno]</code>.</p></dd>

    <dt id="arg-force">force<a class="anchor" aria-label="anchor" href="#arg-force"></a></dt>
<dd><p>If <code>TRUE</code>, INFO fields header and data are overwritten.</p>
<p>If <code>FALSE</code>, an error is thrown if any field already exists.</p></dd>

</dl></div>
    <div id="details">
    <h2>Details</h2>
    <p>Genotypes and frequencies are calculated within the groups of samples
associated with the given level of the given phenotype,
and stored in INFO fields
named according to settings stored in
<code>metadata(vcf)[["TVTBparam"]]</code> (see below).</p>
<p>The <code><a href="TVTBParam-class.html">TVTBparam</a></code> object controls
the key suffix of INFO fields as follows:</p><dl><dt><code>names(ref(param))</code></dt>
<dd><p>Count of reference homozygote genotypes.</p></dd>

    <dt><code>names(het(param))</code></dt>
<dd><p>Count of heterozygote genotypes.</p></dd>

    <dt><code>names(alt(param))</code></dt>
<dd><p>Count of alternate homozygote genotypes.</p></dd>

    <dt><code>aaf(param)</code></dt>
<dd><p>Alternate allele frequency.</p></dd>

    <dt><code>maf(param)</code></dt>
<dd><p>Minor allele frequency</p></dd>


</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p><code>ExpandedVCF</code> object including additional
INFO fields for genotype counts and allele frequencies.
See <em>Details</em>.</p>
    </div>
    <div id="warning">
    <h2>Warning</h2>
    <p>A <code>warning</code> message is issued if genotypes are not fully defined in the
<code>TVTBparam</code>.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Kevin Rue-Albrecht</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="addFrequencies-methods.html">addFrequencies,ExpandedVCF,list-method</a></code>,
<code><a href="addOverallFrequencies-methods.html">addOverallFrequencies,ExpandedVCF-method</a></code>,
<code>VCF</code>,
and <code><a href="TVTBParam-class.html">TVTBparam</a></code>.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Example data ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># VCF file</span></span></span>
<span class="r-in"><span><span class="va">vcfFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"moderate.vcf"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Phenotype file</span></span></span>
<span class="r-in"><span><span class="va">phenoFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"moderate_pheno.txt"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">phenotypes</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">phenoFile</span>, <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># TVTB parameters</span></span></span>
<span class="r-in"><span><span class="va">tparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="TVTBParam-class.html">TVTBparam</a></span><span class="op">(</span><span class="fu"><a href="Genotypes-class.html">Genotypes</a></span><span class="op">(</span><span class="st">"0|0"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0|1"</span>, <span class="st">"1|0"</span><span class="op">)</span>, <span class="st">"1|1"</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Pre-process variants</span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="va">vcfFile</span>, param <span class="op">=</span> <span class="va">tparam</span>, colData <span class="op">=</span> <span class="va">phenotypes</span><span class="op">)</span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu">expand</span><span class="op">(</span><span class="va">vcf</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Example usage ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">addPhenoLevelFrequencies</span><span class="op">(</span><span class="va">vcf</span>, <span class="st">"super_pop"</span>, <span class="st">"AFR"</span><span class="op">)</span></span></span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer></div>






  </body></html>

