<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Genotypes class objects — Genotypes-class • TVTB</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Genotypes class objects — Genotypes-class"><meta property="og:description" content="The Genotypes class stores genotype definitions in a convenient format."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Genotypes class objects</h1>

    <div class="hidden name"><code>Genotypes-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>Genotypes</code> class stores genotype definitions in a convenient format.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">Genotypes</span><span class="op">(</span></span>
<span>    ref <span class="op">=</span> <span class="cn">NA_character_</span>, het <span class="op">=</span> <span class="cn">NA_character_</span>, alt <span class="op">=</span> <span class="cn">NA_character_</span>,</span>
<span>    suffix <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span>ref<span class="op">=</span><span class="st">"REF"</span>, het<span class="op">=</span><span class="st">"HET"</span>, alt<span class="op">=</span><span class="st">"ALT"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="accessor-methods">
    <h2>Accessor methods</h2>
    <p>In the following code snippets <code>x</code> is a <code>Genotypes</code> object.</p><dl><dt><code>ref(x)</code>, <code>ref(x) &lt;- value</code></dt>
<dd><p>Gets or sets the vector that declares homozygote reference genotypes.</p></dd>

    <dt><code>het(x)</code>, <code>het(x) &lt;- value</code></dt>
<dd><p>Gets or sets the vector that declares heterozygote genotypes.</p></dd>

    <dt><code>alt(x)</code>, <code>alt(x) &lt;- value</code></dt>
<dd><p>Gets or sets the vector that declares homozygote alternate genotypes.</p></dd>

    <dt><code>genos(x)</code></dt>
<dd><p>Gets a vector of concatenated
    homozygote reference, heterozygote, and homozygote alternate genotypes.
    See also <code>ref</code>, <code>het</code>, <code>alt</code>, and <code>carrier</code>
    accessors.</p></dd>

    <dt><code>carrier(x)</code></dt>
<dd><p>Gets a vector of concatenated
    heterozygote and homozygote alternate genotypes.
    See also <code>het</code> and <code>alt</code> accessors.</p></dd>

    <dt><code>suffix(x)</code></dt>
<dd><p>Gets a named <code>character</code> vector that declares individual suffixes
    used to store the data for each set of genotypes in the INFO field of the
    <code>VCF</code> object.
    Names of this vector are <code>ref</code>, <code>het</code>, and <code>alt</code>.</p></dd>


</dl></div>
    <div id="arguments">
    <h2>Arguments</h2>
    <p></p>
<dl><dt id="arg-ref">ref<a class="anchor" aria-label="anchor" href="#arg-ref"></a></dt>
<dd><p>A <code>character</code> vector declaring the encoding of
homozygote reference genotypes.</p></dd>

    <dt id="arg-het">het<a class="anchor" aria-label="anchor" href="#arg-het"></a></dt>
<dd><p>A <code>character</code> vector declaring the encoding of
heterozygote genotypes.</p></dd>

    <dt id="arg-alt">alt<a class="anchor" aria-label="anchor" href="#arg-alt"></a></dt>
<dd><p>A <code>character</code> vector declaring the encoding of
homozygote alternate genotypes.</p></dd>

    <dt id="arg-suffix">suffix<a class="anchor" aria-label="anchor" href="#arg-suffix"></a></dt>
<dd><p>Set the individual INFO key suffixes used to
store the statistics of
homozygote reference, heterozygote, and homozygote alternate genotypes,
in this order.
See <em>Details</em> section.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>Genotypes</code> object that contains genotype definitions.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>Genotypes may be initialised as <code>NA_character_</code> and updated from an
imported <code>VCF</code> object using the
<code><a href="autodetectGenotypes.html">autodetectGenotypes</a></code> method.
This may be useful if genotype encodings are not known beforehand.</p>
<p>For each <em>suffix</em> stored in the <code>Genotypes</code> object, <code>TVTB</code>
may store data in the <code>VCF</code> object under the INFO keys defined as follows:</p><dl><dt><em>suffix</em></dt>
<dd><p>Statistics across all samples in the <code>ExpandedVCF</code>
    (<em>e.g.</em> "MAF").</p></dd>

    <dt><em>phenotype_level_suffix</em></dt>
<dd><p>Statistics across samples associated with a given level of a given
    phenotype (<em>e.g.</em> "gender_male_MAF").</p></dd>


</dl><p>Users are recommended to avoid using those INFO keys for other purposes.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Kevin Rue-Albrecht</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code>VCF</code>,
<code><a href="TVTBParam-class.html">TVTBparam</a></code>,
and <code><a href="addCountGenos-methods.html">addCountGenos-methods</a></code>.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Constructors ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">genotypes</span> <span class="op">&lt;-</span> <span class="fu">Genotypes</span><span class="op">(</span><span class="st">"0|0"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0|1"</span>, <span class="st">"1|0"</span><span class="op">)</span>, <span class="st">"1|1"</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Accessors ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Concatenated homozygote reference, heterozygote, and alternate heterozygote</span></span></span>
<span class="r-in"><span><span class="co">## genotypes stored in the Genotypes object returned by the genos() accessor.</span></span></span>
<span class="r-in"><span><span class="fu">genos</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|0" "0|1" "1|0" "1|1"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Individual genotypes can be extracted with ref(), het(), alt() accessors.</span></span></span>
<span class="r-in"><span><span class="fu">ref</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|0"</span>
<span class="r-in"><span><span class="fu">het</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|1" "1|0"</span>
<span class="r-in"><span><span class="fu">alt</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1|1"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Their individual INFO key suffixes can be extracted with suffix() accessors</span></span></span>
<span class="r-in"><span><span class="co">## and the relevant name</span></span></span>
<span class="r-in"><span><span class="fu">suffix</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ref   het   alt </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "REF" "HET" "ALT" </span>
<span class="r-in"><span><span class="fu">suffix</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">[</span><span class="st">"ref"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ref </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "REF" </span>
<span class="r-in"><span><span class="fu">suffix</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">[</span><span class="st">"het"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   het </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "HET" </span>
<span class="r-in"><span><span class="fu">suffix</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">[</span><span class="st">"alt"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   alt </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "ALT" </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Concatenated heterozygote, and alternate heterozygote genotypes are</span></span></span>
<span class="r-in"><span><span class="co">## returned by the carrier() accessor.</span></span></span>
<span class="r-in"><span><span class="fu">carrier</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|1" "1|0" "1|1"</span>
<span class="r-in"><span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="fu">carrier</span><span class="op">(</span><span class="va">genotypes</span><span class="op">)</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> NULL</span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer></div>






  </body></html>

