<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>TVTBparam class objects — TVTBparam-class • TVTB</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="TVTBparam class objects — TVTBparam-class"><meta property="og:description" content="The TVTBparam class stores recurrent parameters of the TVTB
package in a convenient format."><meta property="og:image" content="/logo.png"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">


    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu"><li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul></li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>TVTBparam class objects</h1>

    <div class="hidden name"><code>TVTBParam-class.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>The <code>TVTBparam</code> class stores recurrent parameters of the <code>TVTB</code>
package in a convenient format.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">TVTBparam</span><span class="op">(</span></span>
<span>        <span class="va">genos</span>, ranges <span class="op">=</span> <span class="fu">GRangesList</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        aaf <span class="op">=</span> <span class="st">"AAF"</span>, maf <span class="op">=</span> <span class="st">"MAF"</span>, vep <span class="op">=</span> <span class="st">"CSQ"</span>, bp <span class="op">=</span> <span class="fu">SerialParam</span><span class="op">(</span><span class="op">)</span>,</span>
<span>        svp <span class="op">=</span> <span class="fu">ScanVcfParam</span><span class="op">(</span>which <span class="op">=</span> <span class="fu">reduce</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">ranges</span><span class="op">)</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="accessor-methods">
    <h2>Accessor methods</h2>
    <p>In the following code snippets <code>x</code> is a <code>TVTBparam</code> object.</p><dl><dt><code>genos(x)</code>, <code>genos(x) &lt;- value</code></dt>
<dd><p>Gets or sets the <code>Genotypes</code> object stored in the <code>genos</code> slot.</p></dd>

    <dt><code>ranges(x)</code>, <code>ranges(x) &lt;- value</code></dt>
<dd><p>List of genomic ranges to group variants during analyses and plots.</p></dd>

    <dt><code>ref(x)</code>, <code>ref(x) &lt;- value</code></dt>
<dd><p>Gets or sets the <code>character</code> vector
    that declares homozygote reference genotypes.</p></dd>

    <dt><code>het(x)</code>, <code>het(x) &lt;- value</code></dt>
<dd><p>Gets or sets the <code>character</code> vector
    that declares heterozygote genotypes.</p></dd>

    <dt><code>alt(x)</code>, <code>alt(x) &lt;- value</code></dt>
<dd><p>Gets or sets the <code>character</code> vector
    that declares homozygote alternate genotypes.</p></dd>

    <dt><code>carrier(x)</code></dt>
<dd><p>Gets a <code>character</code> vectors of concatenated
    heterozygote and homozygote alternate genotypes.
    See also <code>het</code> and <code>alt</code> accessors.</p></dd>

    <dt><code>aaf(x)</code>, <code>aaf(x) &lt;- value</code></dt>
<dd><p>Gets or sets the INFO key suffix used to store the
    alternate allele frequency (AAF).</p></dd>

    <dt><code>maf(x)</code>, <code>maf(x) &lt;- value</code></dt>
<dd><p>Gets or sets the INFO key suffix used to store the
    minor allele frequency (MAF).</p></dd>

    <dt><code>vep(x)</code>, <code>maf(x) &lt;- value</code></dt>
<dd><p>Gets or sets the INFO key suffix used to extract the
    VEP predictions.</p></dd>

    <dt><code>bp(x)</code>, <code>bp(x) &lt;- value</code></dt>
<dd><p>Gets or sets the <code>BiocParallel</code> parameters.</p></dd>

    <dt><code>suffix(x)</code></dt>
<dd><p>Gets a named <code>character</code> vector that declares individual suffixes
    used to store the data for each set of genotypes in the INFO field of the
    <code>VCF</code> object.
    Names of this vector are <code>ref</code>, <code>het</code>, <code>alt</code>, <code>aaf</code>,
    and <code>maf</code>.</p></dd>

    <dt><code>svp(x)</code>, <code>svp(x) &lt;- value</code></dt>
<dd><p>Gets or sets the <code>ScanVcfParam</code> parameters.</p></dd>


</dl></div>
    <div id="arguments">
    <h2>Arguments</h2>
    <p></p>
<dl><dt id="arg-genos">genos<a class="anchor" aria-label="anchor" href="#arg-genos"></a></dt>
<dd><p>A <code>Genotypes</code> object that declares the three sets of
homozygote reference, heterozygote, and homozygote alternate genotypes,
as well as the individual key suffix used to
store data for each set of genotypes in the <code>info</code> slot of a
<code>VCF</code> object.
See also <em>Details</em> section.</p></dd>

    <dt id="arg-ranges">ranges<a class="anchor" aria-label="anchor" href="#arg-ranges"></a></dt>
<dd><p>A <code>GRangesList</code> of genomic regions. See <code>svp</code> argument.
<em>In the future, may be used to facet statistics and figures.</em></p></dd>

<dt id="arg-aaf">aaf<a class="anchor" aria-label="anchor" href="#arg-aaf"></a></dt>
<dd><p>INFO key suffix used to store the alternate allele frequency (AAF).</p></dd>

    <dt id="arg-maf">maf<a class="anchor" aria-label="anchor" href="#arg-maf"></a></dt>
<dd><p>INFO key suffix used to store the minor allele frequency (MAF).</p></dd>

    <dt id="arg-vep">vep<a class="anchor" aria-label="anchor" href="#arg-vep"></a></dt>
<dd><p>INFO key suffix used to extract the VEP predictions. See <code>svp</code> argument.</p></dd>

    <dt id="arg-bp">bp<a class="anchor" aria-label="anchor" href="#arg-bp"></a></dt>
<dd><p>A <code>BiocParallelParam</code> object.</p></dd>

    <dt id="arg-svp">svp<a class="anchor" aria-label="anchor" href="#arg-svp"></a></dt>
<dd><p>A <code>ScanVcfParam</code> object.
If none is supplied, the <code>ScanVcfParam</code> slot <code>which</code>
is automatically set to <code>reduce(unlist(ranges))</code>.</p></dd>

</dl></div>
    <div id="value">
    <h2>Value</h2>
    <p>A <code>TVTBparam</code> object that contains recurrent parameters.</p>
    </div>
    <div id="details">
    <h2>Details</h2>
    <p>For each <em>suffix</em> stored in the <code>TVTBparam</code> object, <code>TVTB</code>
may store data in the <code>VCF</code> object under the INFO keys defined as follows:</p><dl><dt><em>suffix</em></dt>
<dd><p>Statistics across all samples in the <code>ExpandedVCF</code>
    (<em>e.g.</em> "MAF").</p></dd>

    <dt><em>phenotype_level_suffix</em></dt>
<dd><p>Statistics across samples associated with a given level of a given
    phenotype (<em>e.g.</em> "gender_male_MAF").</p></dd>


</dl><p>Users are recommended to avoid using those INFO keys for other purposes.</p>
    </div>
    <div id="author">
    <h2>Author</h2>
    <p>Kevin Rue-Albrecht</p>
    </div>
    <div id="see-also">
    <h2>See also</h2>
    <div class="dont-index"><p><code><a href="Genotypes-class.html">Genotypes</a></code>,
<code>VCF</code>,
<code>ExpandedVCF</code>,
<code><a href="addCountGenos-methods.html">addCountGenos-methods</a></code>
<code><a href="vepInPhenoLevel-methods.html">vepInPhenoLevel-methods</a></code>,
<code><a href="variantsInSamples-methods.html">variantsInSamples-methods</a></code>,
and <code>BiocParallelParam</code>.</p></div>
    </div>

    <div id="ref-examples">
    <h2>Examples</h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span class="r-in"><span><span class="co"># Constructors ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">grl</span> <span class="op">&lt;-</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span></span>
<span class="r-in"><span>    <span class="st">"15"</span>, <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span><span class="fl">48413170</span>, <span class="fl">48434757</span>, names <span class="op">=</span> <span class="st">"SLC24A5"</span><span class="op">)</span></span></span>
<span class="r-in"><span>    <span class="op">)</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="va">tparam</span> <span class="op">&lt;-</span> <span class="fu">TVTBparam</span><span class="op">(</span><span class="fu"><a href="Genotypes-class.html">Genotypes</a></span><span class="op">(</span><span class="st">"0|0"</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0|1"</span>, <span class="st">"1|0"</span><span class="op">)</span>, <span class="st">"1|1"</span><span class="op">)</span>, ranges <span class="op">=</span> <span class="va">grl</span><span class="op">)</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co"># Accessors ----</span></span></span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## The Genotypes object stored in the genos slot of the TVTBparam object</span></span></span>
<span class="r-in"><span><span class="co">## return by the genos() accessor.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">genos</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> Class: Genotypes</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   @ref (hom. ref.): "REF" {0|0}</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   @het (heter.)   : "HET" {0|1, 1|0}</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   @alt (hom. alt.): "ALT" {1|1}</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Genomic ranges stored in the TVTBparam object returned by the ranges()</span></span></span>
<span class="r-in"><span><span class="co">## accessor.</span></span></span>
<span class="r-in"><span><span class="fu">ranges</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRangesList object of length 1:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [[1]]</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> GRanges object with 1 range and 0 metadata columns:</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>           seqnames            ranges strand</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>              &lt;Rle&gt;         &lt;IRanges&gt;  &lt;Rle&gt;</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   SLC24A5       15 48413170-48434757      *</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   -------</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   seqinfo: 1 sequence from an unspecified genome; no seqlengths</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Individual genotypes can be extracted with ref(), het(), alt() accessors.</span></span></span>
<span class="r-in"><span><span class="fu">ref</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|0"</span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">het</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|1" "1|0"</span>
<span class="r-in"><span><span class="fu">alt</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "1|1"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Their individual INFO key suffixes can be extracted with suffix() applied to</span></span></span>
<span class="r-in"><span><span class="co">## the above accessors.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ref   het   alt   aaf   maf </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "REF" "HET" "ALT" "AAF" "MAF" </span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">[</span><span class="st">"ref"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   ref </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "REF" </span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">[</span><span class="st">"het"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   het </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "HET" </span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">[</span><span class="st">"alt"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   alt </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "ALT" </span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">[</span><span class="st">"aaf"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   aaf </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "AAF" </span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">suffix</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">[</span><span class="st">"maf"</span><span class="op">]</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   maf </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> "MAF" </span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## Heterozygote, and alternate heterozygote genotypes are</span></span></span>
<span class="r-in"><span><span class="co">## returned by the carrier() accessor.</span></span></span>
<span class="r-in"><span><span class="fu"><a href="Genotypes-class.html">carrier</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "0|1" "1|0" "1|1"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## INFO key suffix of alternate/minor allele frequency returned by the aaf()</span></span></span>
<span class="r-in"><span><span class="co">## and maf() accessors.</span></span></span>
<span class="r-in"><span><span class="fu">aaf</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "AAF"</span>
<span class="r-in"><span><span class="fu">maf</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "MAF"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## INFO key suffix of the VEP predictions returned by the vep() accessor.</span></span></span>
<span class="r-in"><span><span class="fu">vep</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> [1] "CSQ"</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## BiocParallel parameters</span></span></span>
<span class="r-in"><span><span class="fu">bp</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: SerialParam</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bpisup: FALSE; bpnworkers: 1; bptasks: 0; bpjobname: BPJOB</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bplog: FALSE; bpthreshold: INFO; bpstopOnError: TRUE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bpRNGseed: ; bptimeout: NA; bpprogressbar: FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bpexportglobals: FALSE; bpexportvariables: FALSE; bpforceGC: FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bpfallback: FALSE</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bplogdir: NA</span>
<span class="r-out co"><span class="r-pr">#&gt;</span>   bpresultdir: NA</span>
<span class="r-in"><span></span></span>
<span class="r-in"><span><span class="co">## ScanVcfParam parameters</span></span></span>
<span class="r-in"><span><span class="fu">svp</span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></span>
<span class="r-out co"><span class="r-pr">#&gt;</span> class: ScanVcfParam </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> vcfWhich: 1 elements</span>
<span class="r-out co"><span class="r-pr">#&gt;</span> vcfFixed: character() [All] </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> vcfInfo:  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> vcfGeno:  </span>
<span class="r-out co"><span class="r-pr">#&gt;</span> vcfSamples:  </span>
</code></pre></div>
    </div>
  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer></div>






  </body></html>

