<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Introduction to TVTB • TVTB</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Introduction to TVTB">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.2</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Introduction to TVTB</h1>
                        <h4 data-toc-skip class="author">Kévin
Rue-Albrecht</h4>
            <address class="author_afil">
      Department of Medicine, Imperial College London, UKNuffield
Department of Medicine, University of Oxford,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">5 July 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kevinrue/TVTB/blob/devel/vignettes/Introduction.Rmd" class="external-link"><code>vignettes/Introduction.Rmd</code></a></small>
      <div class="hidden name"><code>Introduction.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      The VCF Tool Box (<em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>)
      provides S4 classes and methods to filter, summarise and visualise
      genetic variation data stored in VCF files pre-processed by the
      Ensembl Variant Effect Predictor (VEP). In particular, the package
      extends the <em>FilterRules</em> class (<em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
      package) to define news classes of filter rules applicable to the
      various slots of <code>VCF</code> objects. A Shiny
      web-application, the Shiny Variant Explorer (<em>tSVE</em>),
      provides a convenient interface to demonstrate those
      functionalities integrated in a programming-free environment.
    </div>
    
<div class="section level2">
<h2 id="Introduction">Introduction<a class="anchor" aria-label="anchor" href="#Introduction"></a>
</h2>
<p>The VCF Tool Box (<em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>) offers
S4 classes and methods to filter, summarise and visualise genetic
variation data stored in VCF files pre-processed by the <a href="http://www.ensembl.org/info/docs/tools/vep" class="external-link">Ensembl Variant Effect
Predictor</a> (VEP) <span class="citation">(McLaren et al. 2010)</span>.
An <a href="http://shiny.rstudio.com" class="external-link">RStudio/Shiny</a> web-application,
the Shiny Variant Explorer (<em>tSVE</em>), provides a convenient
interface to demonstrate those functionalities integrated in a
programming-free environment.</p>
<p>Currently, major functionalities in the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> package
include:</p>
<p><strong>A class to store recurrent parameters of genetic
analyses</strong></p>
<ul>
<li>List of reference homozygote, heterozygote, and alternate homozygote
genotype encodings</li>
<li>Key of the INFO field where Ensembl VEP predictions are stored in
the VCF file</li>
<li>Suffix of the INFO fields where calculated data will be stored</li>
<li>List of genomic ranges to analyse and visualise</li>
<li>Parameters for parallel calculations (using <em><a href="https://bioconductor.org/packages/3.20/BiocParallel" class="external-link">BiocParallel</a></em>)</li>
</ul>
<p><strong>Genotype counts and allele frequencies</strong></p>
<ul>
<li>Calculated from the data of an <code>ExpandedVCF</code> objects
(<em>i.e.</em> bi-allelic records)</li>
<li>Stored in INFO fields defined by the above suffixes
<ul>
<li>overall counts and frequencies (<em>i.e.</em> across all
samples)</li>
<li>counts and frequencies within level(s) of phenotype(s)</li>
</ul>
</li>
</ul>
<p><strong>Classes of VCF filter rules</strong></p>
<ul>
<li>Filter rules applicable to the <code>fixed</code> slot of an
<code>VCF</code> object</li>
<li>Filter rules applicable to the <code>info</code> slot of an
<code>VCF</code> object</li>
<li>Filter rules applicable to Ensembl VEP predictions stored in a given
INFO field</li>
<li>A container for combinations of filter rules listed above</li>
<li>Subset <code>VCF</code> objects using the above filter rules</li>
</ul>
</div>
<div class="section level2">
<h2 id="Installation">Installation<a class="anchor" aria-label="anchor" href="#Installation"></a>
</h2>
<p>Instructions to install the VCF Tool Box are available <a href="http://bioconductor.org/packages/TVTB/" class="external-link">here</a>.</p>
<p>Once installed, the package can be loaded and attached as
follows:</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinrue/TVTB" class="external-link">TVTB</a></span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="TVTBparam">Recurrent settings: TVTBparam<a class="anchor" aria-label="anchor" href="#TVTBparam"></a>
</h2>
<p>Most functionalities in <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> require
recurrent information such as:</p>
<ul>
<li>Genotype encoding (homozygote reference, heterozygote, homozygote
alternate),</li>
<li>INFO key that contains the Ensembl VEP predictions in the VCF
file,</li>
<li>List of genomic ranges within which data must be summarised or
visualised
<ul>
<li>affecting the genomic ranges to import from the VCF file</li>
</ul>
</li>
<li>Suffixes of INFO keys where counts and frequencies of genotypes must
be stored:
<ul>
<li>Counts and frequencies may be calculated for individual levels of
selected phenotypes, in which case the data will be stored under INFO
keys formed as
<code>&lt;phenotype&gt;_&lt;level&gt;_&lt;suffix&gt;</code>,</li>
<li>Counts and frequencies across all samples are stored in INFO keys
simply named <code>&lt;suffix&gt;</code>.</li>
</ul>
</li>
<li>Settings for parallel calculations.</li>
</ul>
<p>To reduce the burden of repetition during programming, and to
facilitate analyses using consistent sets of parameters, <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
implements the <code>TVTBparam</code> class. The <code>TVTBparam</code>
class offer a container for parameters recurrently used across the
package. A <code>TVTBparam</code> object may be initialised as
follows:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tparam</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TVTBParam-class.html">TVTBparam</a></span><span class="op">(</span><span class="fu"><a href="../reference/Genotypes-class.html">Genotypes</a></span><span class="op">(</span></span>
<span>    ref <span class="op">=</span> <span class="st">"0|0"</span>,</span>
<span>    het <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"0|1"</span>, <span class="st">"1|0"</span>, <span class="st">"0|2"</span>, <span class="st">"2|0"</span>, <span class="st">"1|2"</span>, <span class="st">"2|1"</span><span class="op">)</span>,</span>
<span>    alt <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"1|1"</span>, <span class="st">"2|2"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    ranges <span class="op">=</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRangesList-class.html" class="external-link">GRangesList</a></span><span class="op">(</span></span>
<span>        SLC24A5 <span class="op">=</span> <span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/GRanges-class.html" class="external-link">GRanges</a></span><span class="op">(</span></span>
<span>            seqnames <span class="op">=</span> <span class="st">"15"</span>,</span>
<span>            <span class="fu">IRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/IRanges/man/IRanges-constructor.html" class="external-link">IRanges</a></span><span class="op">(</span></span>
<span>                start <span class="op">=</span> <span class="fl">48413170</span>, end <span class="op">=</span> <span class="fl">48434757</span><span class="op">)</span></span>
<span>            <span class="op">)</span></span>
<span>        <span class="op">)</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><code>TVTBparam</code> objects have a convenient summary view and
accessor methods:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tparam</span></span></code></pre></div>
<pre><code><span><span class="co">## class: TVTBparam</span></span>
<span><span class="co">## @genos: class: Genotypes</span></span>
<span><span class="co">##     @ref (hom. ref.): "REF" {0|0}</span></span>
<span><span class="co">##     @het (heter.): "HET" {0|1, 1|0, 0|2, 2|0, 1|2, 2|1}</span></span>
<span><span class="co">##     @alt (hom. alt.): "ALT" {1|1, 2|2}</span></span>
<span><span class="co">##   @ranges: 1 GRanges on 1 sequence(s)</span></span>
<span><span class="co">##   @aaf (alt. allele freq.): "AAF"</span></span>
<span><span class="co">##   @maf (minor allele freq.): "MAF"</span></span>
<span><span class="co">##   @vep (Ensembl VEP key): "CSQ"</span></span>
<span><span class="co">##   @svp: &lt;ScanVcfParam object&gt;</span></span>
<span><span class="co">##   @bp: &lt;SerialParam object&gt;</span></span></code></pre>
<p>In this example:</p>
<ul>
<li>
<strong>Genotypes</strong>
<ul>
<li>Accessor: <code>genos(x)</code>
</li>
<li>Class: <code>Genotypes</code>
</li>
<li>Homozygote reference genotype is encoded <code>"0|0"</code>.</li>
<li>Counts of reference genotypes are stored in INFO keys suffixed with
<code>"REF"</code>.</li>
<li>Heterozygote genotypes are encoded as <code>"0|1"</code>,
<code>"1|0"</code>, <code>"0|2"</code>, <code>"2|0"</code>,
<code>"1|2"</code>, and <code>"2|1"</code>.</li>
<li>Counts of heterozygote genotypes are stored in INFO keys suffixed
with <code>"HET"</code>.</li>
<li>Homozygote alternate genotype is encoded <code>"1|1"</code>.</li>
<li>Counts of alternate genotypes are stored in INFO keys suffixed with
<code>"ALT"</code>.</li>
</ul>
</li>
<li>
<strong>Genomic ranges</strong>
<ul>
<li>Accessor: <code>ranges(x)</code>
</li>
<li>Class: <code>GRangesList</code>
</li>
<li>A gene-coding region on chromosome <code>"15"</code>.</li>
</ul>
</li>
<li>
<strong>Alternate allele frequency</strong>
<ul>
<li>Accessor: <code>aaf(x)</code>
</li>
<li>Calculated values will be stored under INFO keys suffixed with
<code>"AAF"</code>.</li>
</ul>
</li>
<li>
<strong>Minor allele frequency</strong>
<ul>
<li>Accessor: <code>maf(x)</code>
</li>
<li>Calculated values will be stored under INFO keys suffixed with
<code>"MAF"</code>.</li>
</ul>
</li>
<li>
<strong>Ensembl VEP predictions</strong>
<ul>
<li>Accessor: <code>vep(x)</code>
</li>
<li>Information will be imported from the INFO field
<code>"CSQ"</code>.</li>
</ul>
</li>
<li>
<strong>Parallel calculations</strong>
<ul>
<li>Accessor: <code>bp(x)</code>
</li>
<li>Class: <code>BiocParallelParam</code>
</li>
<li>Serial evaluation (<em>i.e.</em> do not run parallel code)</li>
</ul>
</li>
<li>
<strong>VCF scan parameters</strong>
<ul>
<li>Accessor: <code>svp(x)</code>
</li>
<li>Class: <code>ScanVcfParam</code>
</li>
<li>
<code>which</code> slot automatically populated with
<code>reduce(unlist(ranges(x)))</code>
</li>
</ul>
</li>
</ul>
<p>Default values are provided for all slots except genotypes, as these
may vary more frequently from one data set to another (<em>e.g.</em>
phased, unphased, imputed).</p>
</div>
<div class="section level2">
<h2 id="Import">Data import<a class="anchor" aria-label="anchor" href="#Import"></a>
</h2>
<div class="section level3">
<h3 id="ImportVariants">Genetic variants<a class="anchor" aria-label="anchor" href="#ImportVariants"></a>
</h3>
<p>Functionalities in <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> support
<code>CollapsedVCF</code> and <code>ExpandedVCF</code> objects (both
extending the virtual class <code>VCF</code>) of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
package.</p>
<p>Typically, <code>CollapsedVCF</code> objects are produced by the
<em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
<code>readVcf</code> method after parsing a VCF file, and
<code>ExpandedVCF</code> objects result of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
<code>expand</code> method applied to a <code>CollapsedVCF</code>
object.</p>
<p>Any information that users deem relevant for the analysis may be
imported from VCF files and stored in <code>VCF</code> objects passed to
<em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
methods. However, to enable the key functionalities of the package, the
slots of a <code>VCF</code> object should include at least the following
information:</p>
<ul>
<li>
<code>fixed(x)</code>
<ul>
<li>fields <code>"REF"</code> and <code>"ALT"</code>.</li>
</ul>
</li>
<li>
<code>info(x)</code>
<ul>
<li>field <code>&lt;vep&gt;</code>: where <code>&lt;vep&gt;</code>
stands for the INFO key where the Ensembl VEP predictions are stored in
the <code>VCF</code> object.</li>
</ul>
</li>
<li>
<code>geno(x)</code>
<ul>
<li>
<code>GT</code>: genotypes.</li>
</ul>
</li>
<li>
<code>colData(x)</code>: phenotypes.</li>
</ul>
</div>
<div class="section level3">
<h3 id="ImportGRangesList">List of genomic ranges<a class="anchor" aria-label="anchor" href="#ImportGRangesList"></a>
</h3>
<p>In the near future, <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
functionalities are expected to produce summary statistics and plots
faceted by meta-features, each potentially composed of multiple genomic
ranges.</p>
<p>For instance, burden tests may be performed on a set of transcripts,
considering only variants in their respective sets of exons. The <em><a href="https://bioconductor.org/packages/3.20/GenomicRanges" class="external-link">GenomicRanges</a></em>
<code>GRangesList</code> class is an ideal container in example, as each
<code>GRanges</code> in the <code>GRangesList</code> would represent a
transcript, and each element in the <code>GRanges</code> would represent
an exon.</p>
<p>Furthermore, <code>TVTBparam</code> objects may be supplied as the
<code>param</code> argument of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
<code>readVcf</code>. In this case, the <code>TVTBparam</code> object is
used to import only variants overlapping the relevant genomic regions.
Moreover, the <code>readVcf</code> method also ensured that the
<code>vep</code> slot of the <code>TVTBparam</code> object is present in
the header of the VCF file.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">svp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/methods/as.html" class="external-link">as</a></span><span class="op">(</span><span class="va">tparam</span>, <span class="st">"ScanVcfParam"</span><span class="op">)</span></span>
<span><span class="va">svp</span></span></code></pre></div>
<pre><code><span><span class="co">## class: ScanVcfParam </span></span>
<span><span class="co">## vcfWhich: 1 elements</span></span>
<span><span class="co">## vcfFixed: character() [All] </span></span>
<span><span class="co">## vcfInfo:  </span></span>
<span><span class="co">## vcfGeno:  </span></span>
<span><span class="co">## vcfSamples:</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="ImportPhenotypes">Phenotypes<a class="anchor" aria-label="anchor" href="#ImportPhenotypes"></a>
</h3>
<p>Although <code>VCF</code> objects may be constructed without attached
phenotype data, phenotype information is critical to interpret and
compare genetic variants between groups of samples (<em>e.g.</em> burden
of damaging variants in different phenotype levels).</p>
<p><code>VCF</code> objects accept phenotype information (as <em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
<code>DataFrame</code>) in the <code>colData</code> slot. This practice
has the key advantage of keeping phenotype and genetic information
synchronised through operation such as subsetting and re-ordering,
limiting workspace <a href="https://en.wikipedia.org/wiki/Software_entropy" class="external-link">entropy</a> and
confusion.</p>
</div>
<div class="section level3">
<h3 id="ImportExample">Example<a class="anchor" aria-label="anchor" href="#ImportExample"></a>
</h3>
<p>An <code>ExpandedVCF</code> object that contains the minimal data
necessary for the rest of the vignette can be created as follows:</p>
<p><strong>Step 1:</strong> Import phenotypes</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phenoFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>, <span class="st">"integrated_samples.txt"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span>
<span><span class="va">phenotypes</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/DataFrame-class.html" class="external-link">DataFrame</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">phenoFile</span>, header <span class="op">=</span> <span class="cn">TRUE</span>, row.names <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 2:</strong> Define the VCF file to parse</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcfFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span></span>
<span>    <span class="st">"extdata"</span>, <span class="st">"chr15.phase3_integrated.vcf.gz"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span>
<span><span class="va">tabixVcf</span> <span class="op">&lt;-</span> <span class="fu">Rsamtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/TabixFile-class.html" class="external-link">TabixFile</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">vcfFile</span><span class="op">)</span></span></code></pre></div>
<p><strong>Step 3:</strong> Define VCF import parameters</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/ScanVcfParam-class.html" class="external-link">vcfInfo</a></span><span class="op">(</span><span class="fu"><a href="../reference/TVTBParam-class.html">svp</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/TVTBParam-class.html">vep</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span>
<span><span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/ScanVcfParam-class.html" class="external-link">vcfGeno</a></span><span class="op">(</span><span class="fu"><a href="../reference/TVTBParam-class.html">svp</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="st">"GT"</span></span>
<span><span class="fu"><a href="../reference/TVTBParam-class.html">svp</a></span><span class="op">(</span><span class="va">tparam</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## class: ScanVcfParam </span></span>
<span><span class="co">## vcfWhich: 1 elements</span></span>
<span><span class="co">## vcfFixed: character() [All] </span></span>
<span><span class="co">## vcfInfo: CSQ </span></span>
<span><span class="co">## vcfGeno: GT </span></span>
<span><span class="co">## vcfSamples:</span></span></code></pre>
<p>Of particular interest in the above chunk of code:</p>
<ul>
<li>The <code>TVTBparam</code> constructor previously populated the
<code>which</code> slot of <code>svp</code> with “reduced”
(<em>i.e.</em> non-overlapping) genomic ranges defined in the
<code>ranges</code> slot.</li>
<li>Only the INFO key defined in the <code>vep</code> slot will be
imported</li>
<li>Only the matrix of called genotypes will be imported</li>
</ul>
<p><strong>Step 4:</strong> Import and pre-process variants</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Import variants as a CollapsedVCF object</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span></span>
<span>    <span class="va">tabixVcf</span>, param <span class="op">=</span> <span class="va">tparam</span>, colData <span class="op">=</span> <span class="va">phenotypes</span><span class="op">)</span></span>
<span><span class="co"># Expand into a ExpandedVCF object (bi-allelic records)</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu">expand</span><span class="op">(</span>x <span class="op">=</span> <span class="va">vcf</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<p>Of particular interest in the above chunk of code, the
<code>readVcf</code> method is given:</p>
<ul>
<li>
<code>TVTBparam</code> parameters, invoking the corresponding method
signature</li>
<li>phenotypes
<ul>
<li>The <code>rownames</code> of those phenotypes defines the sample
identifiers that are queried from the VCF file.</li>
<li>Those phenotypes are stored in the <code>colData</code> slot of the
resulting <code>VCF</code> object.</li>
</ul>
</li>
</ul>
<p>The result is an <code>ExpandedVCF</code> object that includes
variants in the targeted genomic range(s) and samples:</p>
<pre><code><span><span class="co">## class: ExpandedVCF </span></span>
<span><span class="co">## dim: 481 2504 </span></span>
<span><span class="co">## rowRanges(vcf):</span></span>
<span><span class="co">##   GRanges with 5 metadata columns: paramRangeID, REF, ALT, QUAL, FILTER</span></span>
<span><span class="co">## info(vcf):</span></span>
<span><span class="co">##   DataFrame with 1 column: CSQ</span></span>
<span><span class="co">## info(header(vcf)):</span></span>
<span><span class="co">##        Number Type   Description                                               </span></span>
<span><span class="co">##    CSQ .      String Consequence annotations from Ensembl VEP. Format: Allel...</span></span>
<span><span class="co">## geno(vcf):</span></span>
<span><span class="co">##   List of length 1: GT</span></span>
<span><span class="co">## geno(header(vcf)):</span></span>
<span><span class="co">##       Number Type   Description</span></span>
<span><span class="co">##    GT 1      String Genotype</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="Frequencies">Adding allele frequencies<a class="anchor" aria-label="anchor" href="#Frequencies"></a>
</h2>
<p>Although interesting figures and summary tables may be obtained as
soon as the first <code>ExpandedVCF</code> object is created (see
section <a href="#Summarise">Summarising Ensembl VEP predictions</a>),
those methods may benefit from information added to additional INFO keys
after data import, either manually by the user, or through various
methods implemented in the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
package.</p>
<div class="section level3">
<h3 id="FrequenciesOverall">Adding overall frequencies<a class="anchor" aria-label="anchor" href="#FrequenciesOverall"></a>
</h3>
<p>For instance, the method <code>addOverallFrequencies</code> uses the
reference homozoygote (<em>REF</em>), heterozygote (<em>HET</em>), and
homozygote alternate (<em>ALT</em>) genotypes defined in the
<code>TVTBparam</code> object stored in the <code>VCF</code> metadata to
obtain the count of each genotype in an <code>ExpandedVCF</code> object.
Immediately thereafter, the method uses those counts to calculate
alternate allele frequency (<em>AAF</em>) and minor allele frequency
(<em>MAF</em>). Finally, the method stores the five calculated values
(<em>REF</em>, <em>HET</em>, <em>ALT</em>, <em>AAF</em>, and
<em>MAF</em>) in INFO keys defined by suffixes also declared in the
<code>TVTBparam</code> object.</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">initialInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addOverallFrequencies-methods.html">addOverallFrequencies</a></span><span class="op">(</span>vcf <span class="op">=</span> <span class="va">vcf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span>, <span class="va">initialInfo</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "REF" "HET" "ALT" "AAF" "MAF"</span></span></code></pre>
<p>Notably, the <code>addOverallFrequencies</code> method is synonym to
the <code>addFrequencies</code> method missing the argument
<code>phenos</code>:</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addFrequencies-methods.html">addFrequencies</a></span><span class="op">(</span>vcf <span class="op">=</span> <span class="va">vcf</span>, force <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="FrequenciesPhenoLevel">Adding frequencies within phenotype level(s)<a class="anchor" aria-label="anchor" href="#FrequenciesPhenoLevel"></a>
</h3>
<p>Similarly, the method <code>addPhenoLevelFrequencies</code> obtains
the count of each genotype in samples associated with given level(s) of
given phenotype(s), and stores the calculated values in INFO keys
defined as <code>&lt;pheno&gt;_&lt;level&gt;_&lt;suffix&gt;</code>, with
suffixes defined in the <code>TVTBparam</code> object stored in the
<code>VCF</code> metadata.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">initialInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addPhenoLevelFrequencies-methods.html">addPhenoLevelFrequencies</a></span><span class="op">(</span></span>
<span>    vcf <span class="op">=</span> <span class="va">vcf</span>, pheno <span class="op">=</span> <span class="st">"super_pop"</span>, level <span class="op">=</span> <span class="st">"AFR"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span>, <span class="va">initialInfo</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "super_pop_AFR_REF" "super_pop_AFR_HET" "super_pop_AFR_ALT"</span></span>
<span><span class="co">## [4] "super_pop_AFR_AAF" "super_pop_AFR_MAF"</span></span></code></pre>
<p>Notably, the <code>addPhenoLevelFrequencies</code> method is synonym
to the <code>addFrequencies</code> method called with the argument
<code>phenos</code> given as a list where <code>names</code> are
phenotypes, and values are <code>character</code> vectors of levels to
process within each phenotype:</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">initialInfo</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addFrequencies-methods.html">addFrequencies</a></span><span class="op">(</span></span>
<span>    <span class="va">vcf</span>,</span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>super_pop <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"EUR"</span>, <span class="st">"SAS"</span>, <span class="st">"EAS"</span>, <span class="st">"AMR"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/sets.html" class="external-link">setdiff</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span>, <span class="va">initialInfo</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "super_pop_EUR_REF" "super_pop_EUR_HET" "super_pop_EUR_ALT"</span></span>
<span><span class="co">##  [4] "super_pop_EUR_AAF" "super_pop_EUR_MAF" "super_pop_SAS_REF"</span></span>
<span><span class="co">##  [7] "super_pop_SAS_HET" "super_pop_SAS_ALT" "super_pop_SAS_AAF"</span></span>
<span><span class="co">## [10] "super_pop_SAS_MAF" "super_pop_EAS_REF" "super_pop_EAS_HET"</span></span>
<span><span class="co">## [13] "super_pop_EAS_ALT" "super_pop_EAS_AAF" "super_pop_EAS_MAF"</span></span>
<span><span class="co">## [16] "super_pop_AMR_REF" "super_pop_AMR_HET" "super_pop_AMR_ALT"</span></span>
<span><span class="co">## [19] "super_pop_AMR_AAF" "super_pop_AMR_MAF"</span></span></code></pre>
<p>In addition, the <code>addFrequencies</code> method can be given a
<code>character</code> vector of phenotypes as the <code>phenos</code>
argument, in which case frequencies are calculated for <em>all</em>
levels of the given phenotypes:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/addFrequencies-methods.html">addFrequencies</a></span><span class="op">(</span><span class="va">vcf</span>, <span class="st">"pop"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"^pop_[[:alpha:]]+_REF"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span>, value <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "pop_GBR_REF" "pop_FIN_REF" "pop_CHS_REF" "pop_PUR_REF" "pop_CDX_REF"</span></span>
<span><span class="co">## [6] "pop_CLM_REF"</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="Filter">Filtering variants<a class="anchor" aria-label="anchor" href="#Filter"></a>
</h2>
<p>Although <code>VCF</code> objects are straightforward to subset using
either indices and row names (as they inherit from the <em><a href="https://bioconductor.org/packages/3.20/SummarizedExperiment" class="external-link">SummarizedExperiment</a></em>
<code>RangedSummarizedExperiment</code> class), users may wish to
identify variants that pass combinations of criteria based on
information in their <code>fixed</code> slot, <code>info</code> slot,
and Ensembl VEP predictions, a non-trivial task due to those pieces of
information being stored in different slots of the <code>VCF</code>
object, and the <em>1:N</em> relationship between variants and
EnsemblVEP predictions.</p>
<div class="section level3">
<h3 id="FilterDefine">Definition of VCF filter rules<a class="anchor" aria-label="anchor" href="#FilterDefine"></a>
</h3>
<p>To facilitate the definition of VCF filter rules, and their
application to <code>VCF</code> objects, <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> extends
the <em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
<code>FilterRules</code> class in four new classes of filter rules:</p>
<table style="width:93%;" class="table">
<caption>Motivation for each of the new classes extending
<code>FilterRules</code>, to define VCF filter rules.</caption>
<colgroup>
<col width="26%">
<col width="66%">
</colgroup>
<thead><tr class="header">
<th align="center">Class</th>
<th align="center">Motivation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><code>VcfFixedRules</code></td>
<td align="center">Filter rules applied to the <code>fixed</code> slot
of a <code>VCF</code> object.</td>
</tr>
<tr class="even">
<td align="center"><code>VcfInfoRules</code></td>
<td align="center">Filter rules applied to the <code>info</code> slot of
a <code>VCF</code> object.</td>
</tr>
<tr class="odd">
<td align="center"><code>VcfVepRules</code></td>
<td align="center">Filter rules applied to the Ensembl VEP predictions
stored in a given INFO key of a <code>VCF</code> object.</td>
</tr>
<tr class="even">
<td align="center"><code>VcfFilterRules</code></td>
<td align="center">Combination of <code>VcfFixedRules</code>,
<code>VcfInfoRules</code>, and <code>VcfVepRules</code> applicable to a
<code>VCF</code> object.</td>
</tr>
</tbody>
</table>
<p>Note that <code>FilterRules</code> objects themselves are applicable
to <code>VCF</code> objects, with two important difference from the
above specialised classes:</p>
<ul>
<li>Expressions must explicitely refer to the <code>VCF</code>
slots</li>
<li>As a consequence, a single expression can refer to fields from
different <code>VCF</code> slots, for instance:</li>
</ul>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/FilterRules-class.html" class="external-link">FilterRules</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    mixed <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">fixed</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span>,<span class="st">"FILTER"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"PASS"</span> <span class="op">&amp;</span></span>
<span>            <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span>,<span class="st">"MAF"</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">0.05</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## FilterRules of length 1</span></span>
<span><span class="co">## names(1): mixed</span></span></code></pre>
<p>Instances of those classes may be initialised as follows:</p>
<p><strong>VcfFixedRules</strong></p>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fixedR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfFixedRules</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    pass <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">FILTER</span> <span class="op">==</span> <span class="st">"PASS"</span><span class="op">)</span>,</span>
<span>    qual <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">QUAL</span> <span class="op">&gt;</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fixedR</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfFixedRules of length 2</span></span>
<span><span class="co">## names(2): pass qual</span></span></code></pre>
<p><strong>VcfInfoRules</strong></p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">infoR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfInfoRules</a></span><span class="op">(</span></span>
<span>    exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        rare <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">MAF</span> <span class="op">&lt;</span> <span class="fl">0.01</span> <span class="op">&amp;</span> <span class="va">MAF</span> <span class="op">&gt;</span> <span class="fl">0</span><span class="op">)</span>,</span>
<span>        common <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">MAF</span> <span class="op">&gt;</span> <span class="fl">0.05</span><span class="op">)</span>,</span>
<span>        mac_ge3 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">HET</span> <span class="op">+</span> <span class="fl">2</span><span class="op">*</span><span class="va">ALT</span> <span class="op">&gt;=</span> <span class="fl">3</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    active <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">TRUE</span>, <span class="cn">TRUE</span>, <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">infoR</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfInfoRules of length 3</span></span>
<span><span class="co">## names(3): rare common mac_ge3</span></span></code></pre>
<p>The above code chunk illustrates useful features of
<code>FilterRules</code>:</p>
<ul>
<li>
<code>FilterRules</code> are initialised in an active state by
default (evaluating an <em>inactive</em> rule returns <code>TRUE</code>
for all items) The <code>active</code> argument may be used to
initialise specific filter rules in an inactive state.</li>
<li>A single rule <code>expression</code> (or <code>function</code>) may
refer to multiple columns of the relevant slot in the <code>VCF</code>
object.</li>
<li>Rules may include calculations, allowing filtering on values not
necessarily stored as such in any slot of the <code>VCF</code>
object.</li>
</ul>
<p><strong>VcfVepRules</strong></p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vepR</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfVepRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    missense <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">Consequence</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"missense_variant"</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    CADD_gt15 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">CADD_PHRED</span> <span class="op">&gt;</span> <span class="fl">15</span><span class="op">)</span></span>
<span>    <span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vepR</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfVepRules of length 2</span></span>
<span><span class="co">## names(2): missense CADD_gt15</span></span></code></pre>
<p><strong>VcfFilterRules</strong></p>
<p><code>VcfFilterRules</code> combine VCF filter rules of different
types in a single object.</p>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vcfRules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VcfFilterRules-class.html">VcfFilterRules</a></span><span class="op">(</span><span class="va">fixedR</span>, <span class="va">infoR</span>, <span class="va">vepR</span><span class="op">)</span></span>
<span><span class="va">vcfRules</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfFilterRules of length 7</span></span>
<span><span class="co">## names(7): pass qual rare common mac_ge3 missense CADD_gt15</span></span></code></pre>
<p>This vignette offers only a brief peek into the utility and
flexibility of VCF filter rules. More (complex) examples are given in a
separate vignette, including filter rules using functions and pattern
matching. The documentation of the <em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
package—where the parent class <code>FilterRules</code> is defined—can
also be a source of inspiration.</p>
</div>
<div class="section level3">
<h3 id="FilterControl">Control of VCF filter rules<a class="anchor" aria-label="anchor" href="#FilterControl"></a>
</h3>
<p>As the above classes of <em>VCF filter rules</em> inherit from the
<em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
<code>FilterRules</code> class, they also benefit from its accessors and
methods. For instance, VCF filter rules can easily be toggled between
active and inactive states:</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">vcfRules</span><span class="op">)</span><span class="op">[</span><span class="st">"CADD_gt15"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span><span class="fu">active</span><span class="op">(</span><span class="va">vcfRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      pass      qual      rare    common   mac_ge3  missense CADD_gt15 </span></span>
<span><span class="co">##      TRUE      TRUE      TRUE      TRUE     FALSE      TRUE     FALSE</span></span></code></pre>
<p>A separate vignette describes in greater detail the use of classes
that contain <em>VCF filter rules</em>.</p>
</div>
<div class="section level3">
<h3 id="FilterEval">Evaluation of VCF filter rules<a class="anchor" aria-label="anchor" href="#FilterEval"></a>
</h3>
<p>Once defined, the above filter rules can be applied to
<code>ExpandedVCF</code> objects, in the same way as
<code>FilterRules</code> are evaluated in a given environment (see the
<em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
documentation):</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span>expr <span class="op">=</span> <span class="va">infoR</span>, envir <span class="op">=</span> <span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE </span></span>
<span><span class="co">## logical     481</span></span></code></pre>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span>expr <span class="op">=</span> <span class="va">vcfRules</span>, envir <span class="op">=</span> <span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE </span></span>
<span><span class="co">## logical     481</span></span></code></pre>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span>expr <span class="op">=</span> <span class="va">vcfRules</span>, envir <span class="op">=</span> <span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    pass           qual            rare           common        mac_ge3       </span></span>
<span><span class="co">##  Mode:logical   Mode:logical   Mode :logical   Mode :logical   Mode:logical  </span></span>
<span><span class="co">##  TRUE:481       TRUE:481       FALSE:45        FALSE:453       TRUE:481      </span></span>
<span><span class="co">##                                TRUE :436       TRUE :28                      </span></span>
<span><span class="co">##   missense       CADD_gt15     </span></span>
<span><span class="co">##  Mode :logical   Mode:logical  </span></span>
<span><span class="co">##  FALSE:454       TRUE:481      </span></span>
<span><span class="co">##  TRUE :27</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="Visualise">Visualising data<a class="anchor" aria-label="anchor" href="#Visualise"></a>
</h2>
<div class="section level3">
<h3 id="visualise-info-data-by-phenotype-level-on-a-genomic-axis">Visualise INFO data by phenotype level on a genomic axis<a class="anchor" aria-label="anchor" href="#visualise-info-data-by-phenotype-level-on-a-genomic-axis"></a>
</h3>
<p>Let us show the alternate allele frequency (AAF) of common variants,
estimated in each super-population, in the context of the transcripts
ovelapping the region of interest.</p>
<p>In the <code>MAF</code> track:</p>
<ul>
<li>the points represent the MAF in each super-population on a common Y
axis.</li>
<li>the heatmap represents the MAF as the color intensity, given a row
for each super-population.</li>
</ul>
<div class="sourceCode" id="cb40"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfo-methods.html">plotInfo</a></span><span class="op">(</span></span>
<span>        <span class="fu">subsetByFilter</span><span class="op">(</span><span class="va">vcf</span>, <span class="va">vcfRules</span><span class="op">[</span><span class="st">"common"</span><span class="op">]</span><span class="op">)</span>, <span class="st">"AAF"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu">EnsDb.Hsapiens.v75</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/EnsDb.Hsapiens.v75/man/package.html" class="external-link">EnsDb.Hsapiens.v75</a></span>,</span>
<span>        <span class="st">"super_pop"</span>,</span>
<span>        zero.rm <span class="op">=</span> <span class="cn">FALSE</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="Introduction_files/figure-html/plotMAFCommon-1.png" width="700"></p>
<p>Alternatively, the minor allele frequency (MAF) of missense variants
(as estimated from the entire data set) may be visualised in the same
manner. However, due to the nature of those variants, the
<code>zero.rm</code> argument may be set to <code>TRUE</code> to hide
all data points showing a MAF of <code>0</code>; thereby variants
actually detected in each super-population are emphasised even at low
frequencies.</p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotInfo-methods.html">plotInfo</a></span><span class="op">(</span></span>
<span>        <span class="fu">subsetByFilter</span><span class="op">(</span><span class="va">vcf</span>, <span class="va">vcfRules</span><span class="op">[</span><span class="st">"missense"</span><span class="op">]</span><span class="op">)</span>, <span class="st">"MAF"</span>,</span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/range.html" class="external-link">range</a></span><span class="op">(</span><span class="fu">GenomicRanges</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/GenomicRanges/man/genomic-range-squeezers.html" class="external-link">granges</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span>,</span>
<span>        <span class="fu">EnsDb.Hsapiens.v75</span><span class="fu">::</span><span class="va"><a href="https://rdrr.io/pkg/EnsDb.Hsapiens.v75/man/package.html" class="external-link">EnsDb.Hsapiens.v75</a></span>,</span>
<span>        <span class="st">"super_pop"</span>,</span>
<span>        zero.rm <span class="op">=</span> <span class="cn">TRUE</span></span>
<span>    <span class="op">)</span></span></code></pre></div>
<p><img src="Introduction_files/figure-html/plotMAFrare-1.png" width="700"></p>
</div>
</div>
<div class="section level2">
<h2 id="pairwise-comparison-of-info-data-between-phenotype-levels">Pairwise comparison of INFO data between phenotype levels<a class="anchor" aria-label="anchor" href="#pairwise-comparison-of-info-data-between-phenotype-levels"></a>
</h2>
<p>Using the <em><a href="https://CRAN.R-project.org/package=GGally" class="external-link">GGally</a></em>
<code>ggpairs</code> method, let us make a matrix of plots for common
variants, showing:</p>
<ul>
<li>in the <em>lower</em> triangle, pairwise scatter plots of the
alternate allele frequency estimated each super-population</li>
<li>on the <em>diagonal</em>, the density plot for each super-population
shown on the diagonal</li>
<li>in the <em>upper</em> triangle, the correlation<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a> value</li>
</ul>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/pairsInfo-methods.html">pairsInfo</a></span><span class="op">(</span><span class="fu">subsetByFilter</span><span class="op">(</span><span class="va">vcf</span>, <span class="va">vcfRules</span><span class="op">[</span><span class="st">"common"</span><span class="op">]</span><span class="op">)</span>, <span class="st">"AAF"</span>, <span class="st">"super_pop"</span><span class="op">)</span></span></code></pre></div>
<p><img src="Introduction_files/figure-html/pairsInfo-1.png" width="700"></p>
<p>Note that the ellipsis <code>...</code> allows a high degree of
customisation, as it passes additional arguments to the underlying
<code>ggpairs</code> method.</p>
</div>
<div class="section level2">
<h2 id="a-taste-of-future">A taste of future…<a class="anchor" aria-label="anchor" href="#a-taste-of-future"></a>
</h2>
<p>This section presents upcoming features.</p>
<div class="section level3">
<h3 id="Summarise">Summarising Ensembl VEP predictions<a class="anchor" aria-label="anchor" href="#Summarise"></a>
</h3>
<p>As soon as genetic and phenotypic information are imported into an
<code>ExpandedVCF</code> object, or after the object was extended with
additional information, the scientific value of the data may be revealed
by a variety of summary statistics and graphical representations. This
section will soon present several ideas <em>being implemented</em> in
<em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>,
for instance:</p>
<ul>
<li>Count of discrete Ensembl VEP predictions
<ul>
<li>by phenotype level</li>
<li>by genomic feature affected (<em>i.e.</em> transcript)</li>
</ul>
</li>
<li>Distribution of continuous Ensembl VEP predictions
<ul>
<li>by phenotype level</li>
<li>by genomic feature affected (<em>i.e.</em> transcript)</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="acknowledgement">Acknowledgement<a class="anchor" aria-label="anchor" href="#acknowledgement"></a>
</h2>
<p>Dr. Stefan Gräf and Mr. Matthias Haimel for advice on the VCF file
format and the Ensembl VEP script. Prof. Martin Wilkins for his trust
and support. Dr. Michael Lawrence for his helpful code review and
suggestions.</p>
<p>Last but not least, the amazing collaborative effort of the
<code>rep("many",n)</code> <a href="http://bioconductor.org" class="external-link">Bioconductor</a> developers whose hard
work appears through the dependencies of this package.</p>
</div>
<div class="section level2">
<h2 id="SessionInfo">Session info<a class="anchor" aria-label="anchor" href="#SessionInfo"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] TVTB_1.31.2      knitr_1.47       BiocStyle_2.33.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          rstudioapi_0.16.0          </span></span>
<span><span class="co">##   [3] jsonlite_1.8.8              magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.57.0      farver_2.1.2               </span></span>
<span><span class="co">##   [7] rmarkdown_2.27              fs_1.6.4                   </span></span>
<span><span class="co">##   [9] BiocIO_1.15.0               zlibbioc_1.51.1            </span></span>
<span><span class="co">##  [11] ragg_1.3.2                  vctrs_0.6.5                </span></span>
<span><span class="co">##  [13] memoise_2.0.1               Rsamtools_2.21.0           </span></span>
<span><span class="co">##  [15] RCurl_1.98-1.14             base64enc_0.1-3            </span></span>
<span><span class="co">##  [17] htmltools_0.5.8.1           S4Arrays_1.5.3             </span></span>
<span><span class="co">##  [19] progress_1.2.3              curl_5.2.1                 </span></span>
<span><span class="co">##  [21] SparseArray_1.5.12          Formula_1.2-5              </span></span>
<span><span class="co">##  [23] sass_0.4.9                  bslib_0.7.0                </span></span>
<span><span class="co">##  [25] htmlwidgets_1.6.4           desc_1.4.3                 </span></span>
<span><span class="co">##  [27] plyr_1.8.9                  Gviz_1.49.0                </span></span>
<span><span class="co">##  [29] httr2_1.0.1                 cachem_1.1.0               </span></span>
<span><span class="co">##  [31] GenomicAlignments_1.41.0    lifecycle_1.0.4            </span></span>
<span><span class="co">##  [33] pkgconfig_2.0.3             Matrix_1.7-0               </span></span>
<span><span class="co">##  [35] R6_2.5.1                    fastmap_1.2.0              </span></span>
<span><span class="co">##  [37] GenomeInfoDbData_1.2.12     MatrixGenerics_1.17.0      </span></span>
<span><span class="co">##  [39] digest_0.6.36               colorspace_2.1-0           </span></span>
<span><span class="co">##  [41] GGally_2.2.1                AnnotationDbi_1.67.0       </span></span>
<span><span class="co">##  [43] S4Vectors_0.43.1            textshaping_0.4.0          </span></span>
<span><span class="co">##  [45] Hmisc_5.1-3                 GenomicRanges_1.57.1       </span></span>
<span><span class="co">##  [47] RSQLite_2.3.7               labeling_0.4.3             </span></span>
<span><span class="co">##  [49] filelock_1.0.3              fansi_1.0.6                </span></span>
<span><span class="co">##  [51] httr_1.4.7                  abind_1.4-5                </span></span>
<span><span class="co">##  [53] compiler_4.4.1              withr_3.0.0                </span></span>
<span><span class="co">##  [55] bit64_4.0.5                 pander_0.6.5               </span></span>
<span><span class="co">##  [57] backports_1.5.0             htmlTable_2.4.2            </span></span>
<span><span class="co">##  [59] BiocParallel_1.39.0         DBI_1.2.3                  </span></span>
<span><span class="co">##  [61] ggstats_0.6.0               highr_0.11                 </span></span>
<span><span class="co">##  [63] biomaRt_2.61.2              rappdirs_0.3.3             </span></span>
<span><span class="co">##  [65] DelayedArray_0.31.5         rjson_0.2.21               </span></span>
<span><span class="co">##  [67] tools_4.4.1                 foreign_0.8-87             </span></span>
<span><span class="co">##  [69] nnet_7.3-19                 glue_1.7.0                 </span></span>
<span><span class="co">##  [71] restfulr_0.0.15             grid_4.4.1                 </span></span>
<span><span class="co">##  [73] checkmate_2.3.1             reshape2_1.4.4             </span></span>
<span><span class="co">##  [75] cluster_2.1.6               generics_0.1.3             </span></span>
<span><span class="co">##  [77] gtable_0.3.5                BSgenome_1.73.0            </span></span>
<span><span class="co">##  [79] ensembldb_2.29.0            tidyr_1.3.1                </span></span>
<span><span class="co">##  [81] data.table_1.15.4           hms_1.1.3                  </span></span>
<span><span class="co">##  [83] xml2_1.3.6                  utf8_1.2.4                 </span></span>
<span><span class="co">##  [85] XVector_0.45.0              BiocGenerics_0.51.0        </span></span>
<span><span class="co">##  [87] pillar_1.9.0                stringr_1.5.1              </span></span>
<span><span class="co">##  [89] limma_3.61.2                dplyr_1.1.4                </span></span>
<span><span class="co">##  [91] BiocFileCache_2.13.0        lattice_0.22-6             </span></span>
<span><span class="co">##  [93] deldir_2.0-4                rtracklayer_1.65.0         </span></span>
<span><span class="co">##  [95] bit_4.0.5                   EnsDb.Hsapiens.v75_2.99.0  </span></span>
<span><span class="co">##  [97] biovizBase_1.53.0           tidyselect_1.2.1           </span></span>
<span><span class="co">##  [99] Biostrings_2.73.1           gridExtra_2.3              </span></span>
<span><span class="co">## [101] bookdown_0.40               ProtGenerics_1.37.0        </span></span>
<span><span class="co">## [103] IRanges_2.39.1              SummarizedExperiment_1.35.1</span></span>
<span><span class="co">## [105] stats4_4.4.1                xfun_0.45                  </span></span>
<span><span class="co">## [107] Biobase_2.65.0              statmod_1.5.0              </span></span>
<span><span class="co">## [109] matrixStats_1.3.0           stringi_1.8.4              </span></span>
<span><span class="co">## [111] UCSC.utils_1.1.0            lazyeval_0.2.2             </span></span>
<span><span class="co">## [113] yaml_2.3.8                  evaluate_0.24.0            </span></span>
<span><span class="co">## [115] codetools_0.2-20            interp_1.1-6               </span></span>
<span><span class="co">## [117] tibble_3.2.1                BiocManager_1.30.23        </span></span>
<span><span class="co">## [119] cli_3.6.3                   rpart_4.1.23               </span></span>
<span><span class="co">## [121] systemfonts_1.1.0           munsell_0.5.1              </span></span>
<span><span class="co">## [123] jquerylib_0.1.4             dichromat_2.0-0.1          </span></span>
<span><span class="co">## [125] Rcpp_1.0.12                 GenomeInfoDb_1.41.1        </span></span>
<span><span class="co">## [127] dbplyr_2.5.0                png_0.1-8                  </span></span>
<span><span class="co">## [129] XML_3.99-0.17               parallel_4.4.1             </span></span>
<span><span class="co">## [131] pkgdown_2.0.9.9000          ggplot2_3.5.1              </span></span>
<span><span class="co">## [133] blob_1.2.4                  prettyunits_1.2.0          </span></span>
<span><span class="co">## [135] jpeg_0.1-10                 latticeExtra_0.6-30        </span></span>
<span><span class="co">## [137] AnnotationFilter_1.29.0     bitops_1.0-7               </span></span>
<span><span class="co">## [139] VariantAnnotation_1.51.0    scales_1.3.0               </span></span>
<span><span class="co">## [141] purrr_1.0.2                 crayon_1.5.3               </span></span>
<span><span class="co">## [143] rlang_1.1.4                 KEGGREST_1.45.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="References">References<a class="anchor" aria-label="anchor" href="#References"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-RN1" class="csl-entry">
McLaren, W., B. Pritchard, D. Rios, Y. Chen, P. Flicek, and F.
Cunningham. 2010. <span>“Deriving the Consequences of Genomic Variants
with the Ensembl <span>API</span> and <span>SNP</span> Effect
Predictor.”</span> Journal Article. <em>Bioinformatics</em> 26 (16):
2069–70. <a href="https://doi.org/10.1093/bioinformatics/btq330" class="external-link">https://doi.org/10.1093/bioinformatics/btq330</a>.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>Pearson, by default<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
