<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>The Shiny Variant Explorer • TVTB</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="The Shiny Variant Explorer">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>The Shiny Variant Explorer</h1>
                        <h4 data-toc-skip class="author">Kévin
Rue-Albrecht</h4>
            <address class="author_afil">
      Department of Medicine, Imperial College London, UKNuffield
Department of Medicine, University of Oxford,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">4 July 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kevinrue/TVTB/blob/devel/vignettes/tSVE.Rmd" class="external-link"><code>vignettes/tSVE.Rmd</code></a></small>
      <div class="hidden name"><code>tSVE.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p><a href="http://shiny.rstudio.com" class="external-link">Shiny</a> web-application
      that demonstrates the functionalities of the <em>TVTB</em> package
      integrated in a programming-free environment.</p>
    </div>
    
<div class="section level2">
<h2 id="Preliminary">Preliminary notes<a class="anchor" aria-label="anchor" href="#Preliminary"></a>
</h2>
<p>The Shiny Variant Explorer (<em>tSVE</em>) was primarily developped
to demonstrate features implemented in the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>,
<strong>not</strong> as a production environment. As a result, a few
important considerations should be made to clarify what should and
should <strong>not</strong> be expected from the web-application:</p>
<ul>
<li>Bug fixes will be treated with a much lower priority relative to
those related to package methods.</li>
<li>It is technically not feasible to offer in a web-interface the same
degree of flexibility as the command-line environment (<em>e.g.</em>
<code>...</code><a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>).</li>
<li>Greater control over the input and output data is possible at the
command-line (<em>e.g.</em> refinement of <code>ggplot</code> objects,
definition of custom genomic ranges).</li>
<li>Requests for new features should apply to the package first. Only
features relevant to the package functionalities may be made available
in the web-application.</li>
<li>Figures (<code>ggplot</code>) are currently the only output that can
be exported from the web-application (using the web browser “Download
image”, or equivalent context menu item). In the future, action buttons
may be added to export tables (<em>e.g.</em> CSV format) and figures
(<em>e.g.</em> PDF format).</li>
<li>This vignette is largely static as the web-application may only be
used in an interactive session.</li>
<li>First-time users are encouraged to follow this vignette sequentially
(<em>i.e</em> in order, without skipping sections), as it takes readers
through the sequence of actions of a typical analysis.
<ul>
<li>This vignette was designed to be read beside an open <em>R</em>
session with the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> package
installed, so that users may follow the instructions marked by the word
<strong>Action</strong> and bulleted points in the following
sections.</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="Prerequisites">Pre-requisites<a class="anchor" aria-label="anchor" href="#Prerequisites"></a>
</h2>
<p>The <em>Shiny Variant Explorer</em> suggests a few additional package
dependencies compared to the package, to support certain forms of data
input and display.</p>
<p><strong>Input</strong></p>
<ul>
<li>The <em><a href="https://bioconductor.org/packages/3.20/ensembldb" class="external-link">ensembldb</a></em>
package and relevant <code>EnsDb</code><a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> annotation packages
are required if that interface is used to query genomic ranges
(demonstrated in this <a href="#InputEnsDb">section</a>).</li>
<li>The <em><a href="https://bioconductor.org/packages/3.20/EnsDb.Hsapiens.v75" class="external-link">EnsDb.Hsapiens.v75</a></em>
is required to query genomic ranges associated by gene names for the
demonstration data<a href="#fn3" class="footnote-ref" id="fnref3"><sup>3</sup></a>.</li>
<li>The <em><a href="https://bioconductor.org/packages/3.20/rtracklayer" class="external-link">rtracklayer</a></em>
package is required if a BED file is used to provide genomic ranges
(demonstrated in this <a href="#InputBED">section</a>).</li>
</ul>
<p><strong>Display</strong></p>
<ul>
<li>The latest version of the <em><a href="https://github.com/rstudio/DT" class="external-link">DT</a></em> package is recommended
to benefit from the latest developments (<em>e.g.</em> column filters
inactivated if a single value exist in that column; version <em>&gt;=
0.2.2</em>).</li>
<li>The <em><a href="https://github.com/rstudio/shiny" class="external-link">shiny</a></em>
package is required for all Shiny web-applications.</li>
</ul>
</div>
<div class="section level2">
<h2 id="Launch">Launching the Shiny Variant Explorer<a class="anchor" aria-label="anchor" href="#Launch"></a>
</h2>
<p>The <code><a href="../reference/tSVE.html">TVTB::tSVE()</a></code> method launches the
web-application.</p>
</div>
<div class="section level2">
<h2 id="OverallLayout">Overall layout of the web-application<a class="anchor" aria-label="anchor" href="#OverallLayout"></a>
</h2>
<p>Overall, the web-application is implemented as a web-page with a top
level navigation bar organised from left to right to reflect progression
through a typical analysis, with the exception of the last two menu
items <strong>Settings</strong> and <strong>Session</strong>, which may
be useful to check and update at any point.</p>
<p>Here is a brief overview of the menu items:</p>
<ul>
<li>
<strong>Input</strong>
<ul>
<li>Control which samples, phenotypes, genomic ranges, and VCF fields
must be imported.</li>
<li>An <code>EnsDb</code> annotation package may be selected to use the
associated database interface.</li>
</ul>
</li>
<li>
<strong>Frequencies</strong>
<ul>
<li>Add and remove INFO fields that contain calculated genotype counts
and allele frequencies.</li>
<li>Add and remove genotype counts and allele frequencies across all
samples, or within individual phenotype levels.</li>
</ul>
</li>
<li>
<strong>Filters</strong>
<ul>
<li>Define and apply <em>VCF filter rules</em> (detailed in a separate
vignette).</li>
</ul>
</li>
<li>
<strong>Views</strong>
<ul>
<li>Display and examine major objects of the analysis and their
slots.</li>
</ul>
</li>
<li>
<strong>Plots</strong>
<ul>
<li>Display data plots and associated data tables.</li>
</ul>
</li>
<li>
<strong>Settings</strong>
<ul>
<li>Control advanced parameters of the analysis and
web-application.</li>
</ul>
</li>
<li>
<strong>Session</strong>
<ul>
<li>Display session information and other relevant information.</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="Input">Input panel<a class="anchor" aria-label="anchor" href="#Input"></a>
</h2>
<p>The <strong>Input</strong> panel controls the major input parameters
of the analysis, including phenotypes (and therefore samples), genomic
ranges, and fields to import from VCF file(s). Those inputs are useful
to import only data of interest, as well as to limit memory usage and
duration of calculations.</p>
<div class="section level3">
<h3 id="InputPhenotypes">Phenotypes<a class="anchor" aria-label="anchor" href="#InputPhenotypes"></a>
</h3>
<p>Phenotypes are critical to define groups of samples that may be
compared in summary statistics, tables, and plots. Moreover, phenotypes
also implicitely define the set of samples required in the analysis
(unique sample identifiers usually set as <code>rownames</code> of the
phenotypes).</p>
<p>The web-application accepts phenotypes stored in a text file, with
the following requirements:</p>
<ul>
<li>Fields must be delimited by “white space” (default separator for the
<code>read.table</code> function).</li>
<li>The first column of the file must contain unique sample identifiers,
as syntactically valid <code>rownames</code>.</li>
<li>The first row of the file must phenotype names, as syntactically
valid <code>colnames</code>.</li>
</ul>
<p>When provided, phenotypes will be used to import from VCF file(s)
only genotypes for the corresponding samples identifiers. Moreover, an
error message will be displayed if any of the sample identifiers present
in the phenotypes is absent from the VCF file(s).</p>
<p>Note that the web-application does not <em>absolutely</em> require
phenotype information. In the absence of phenotype information, all
samples are imported from VCF file(s).</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Click on the <em>Browse</em> action button</li>
<li>Navigate to the <code>extdata</code> folder of the <em>TVTB</em>
installation directory</li>
<li>Select the file <code>integrated_samples.txt</code>
</li>
</ul>
<p>Alternatively: click the <em>Sample file</em> button</p>
</blockquote>
<p><strong>Notes</strong></p>
<ul>
<li>The <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
installation directory can be identified using the following command in
an <em>R</em> session:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span></code></pre></div>
<ul>
<li>The file selection pop-up window that is open by the action button
browses files on the server side. <em>This point is only relevant if the
package/web-application is run on a remote server.</em>
</li>
</ul>
</div>
<div class="section level3">
<h3 id="InputGRanges">Genomic ranges<a class="anchor" aria-label="anchor" href="#InputGRanges"></a>
</h3>
<p>Genomic ranges are critical to import only variants in targeted
genomic regions or features (<em>e.g.</em> genes, transcripts, exons),
as well as to limit memory usage and duration of calculations.</p>
<p>The Shiny Variant Explorer currently supports three types of input to
define genomic ranges:</p>
<ul>
<li>BED file</li>
<li>UCSC-style text input</li>
<li>
<code>EnsDb</code> annotation packages</li>
</ul>
<p>Currently, the web-application uses genomic ranges solely to query
the corresponding variants from VCF file(s). In the future, those
genomic ranges may also be used to produce faceted summary statistics
and plots.</p>
<p><strong>Notes</strong>:</p>
<ul>
<li>The web-application does not <em>absolutely</em> require genomic
ranges. In the absence of genomic ranges, all variants are imported from
VCF file(s). <em>Caution recommended with large files!</em>
</li>
<li>When VCF file(s) are parsed (in a later <a href="#InputVariants">section</a>), only the genomic ranges from the
currently selected input mode are</li>
<li>The active genomic ranges are only taken from the currently selected
input mode</li>
</ul>
<div class="section level4">
<h4 id="InputBED">BED file<a class="anchor" aria-label="anchor" href="#InputBED"></a>
</h4>
<p>If a BED file is supplied, the web-application parses it using the
<em><a href="https://bioconductor.org/packages/3.20/rtracklayer" class="external-link">rtracklayer</a></em>
<code>import.bed</code> method. Therefore the file must respect the <a href="https://genome.ucsc.edu/FAQ/FAQformat.html#format1" class="external-link">BED file
format</a> guidelines.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Click on the <em>Browse</em> action button</li>
<li>Navigate to the <code>extdata</code> folder of the <em>TVTB</em>
installation directory</li>
<li>Select the file <code>SLC24A5.bed</code>
</li>
</ul>
<p>Alternatively: click the <em>Sample file</em> button</p>
</blockquote>
<p><strong>Notes</strong>:</p>
<ul>
<li>The BED file defines the same genomic range that was used to extract
variant from the <a href="http://www.1000genomes.org" class="external-link">1000 Genomes
Project</a> Phase 3 release VCF file used in this vignette.</li>
<li>The file selection window that is open by the action button browses
files on the server side (see <a href="#InputPhenotypes">Phenotypes</a>
section above).</li>
</ul>
</div>
<div class="section level4">
<h4 id="InputUCSC">UCSC format<a class="anchor" aria-label="anchor" href="#InputUCSC"></a>
</h4>
<p>Sequence names (<em>i.e.</em> chromosomes), start, and end positions
of one or more genomic ranges may be defined in the text field, with
individual regions separated by <code>";"</code>.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Paste <code>15:48,413,169-48,434,869</code> in the text field</li>
</ul>
<p>Alternatively: click the <em>Sample input</em> button</p>
</blockquote>
<p><strong>Notes</strong>:</p>
<ul>
<li>The web-application automatically trims <code>","</code> characters
from the text input, before coercing the start and end positions to
<code>numeric</code>
</li>
<li>Multiple genomic ranges may be supplied (<em>e.g.</em>
<code>1:123-456;2:234-345;2:456-789</code>)</li>
</ul>
</div>
<div class="section level4">
<h4 id="InputEnsDb">Ensembl-based annotation packages<a class="anchor" aria-label="anchor" href="#InputEnsDb"></a>
</h4>
<p>Currently, genomic ranges encoding only gene-coding regions may be
retrieved from an Ensembl-based database. This feature was adapted from
the web-application implemented in the <em><a href="https://bioconductor.org/packages/3.20/ensembldb" class="external-link">ensembldb</a></em>
package.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Paste <code>SLC24A5</code> in the text field</li>
</ul>
<p>Alternatively: click the <em>Sample input</em> button</p>
</blockquote>
</div>
</div>
<div class="section level3">
<h3 id="InputVariants">Variants<a class="anchor" aria-label="anchor" href="#InputVariants"></a>
</h3>
<p>At the core of the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
package, variants must be imported from one or more VCF file(s)
annotated by the Ensembl Variant Effect Predictor (<a href="http://www.ensembl.org/info/docs/tools/vep" class="external-link">VEP</a>) script <span class="citation">(McLaren et al. 2010)</span>.</p>
<p>Considering the large size of most VCF file(s), it is common practice
to split genetic variants into multiple files, each file used to store
variants located on a single chromosome (more generally; a single
sequence). The Shiny Variant Explorer supports two situations:</p>
<ul>
<li>All variants are stored in a single VCF file (“Single-VCF”
mode).</li>
<li>Variants are split into one file per sequence, with the requirement
that files be named with a pattern including the sequence name (must
match the <code>seqnames</code> slot of the genomic ranges described <a href="#InputGRanges">above</a> (“Multi-VCF mode”).</li>
</ul>
<p>In addition, VCF files can store a plethora of information in their
various fields. It is often useful to select only a subset of fields
relevant for a particular analysis, to limit memory usage. The
web-application uses the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
<code>scanVcfHeader</code> to parse the header of the VCF file
(<em>Single-VCF</em> mode) or the first VCF file (<em>Multi-VCF</em>
mode), to display the list of available fields that users may choose to
import. A few considerations must be made:</p>
<ul>
<li>The web-application requires that Ensembl VEP predictions be present
in the INFO field.</li>
<li>The web-application requires that the <code>"GT"</code> key be
present in the FORMAT field.</li>
</ul>
<div class="section level4">
<h4 id="SingleVCF">Single-VCF mode<a class="anchor" aria-label="anchor" href="#SingleVCF"></a>
</h4>
<p>This mode display an action button that must be used to select the
VCF file from which to import variants.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Click on the <em>Browse</em> action button</li>
<li>Navigate to the <code>extdata</code> folder of the <em>TVTB</em>
installation directory</li>
<li>Select the file <code>chr15.phase3_integrated.vcf.gz</code>
</li>
</ul>
<p>Alternatively: click the <em>Sample file</em> button</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="MultiVCF">Multi-VCF mode<a class="anchor" aria-label="anchor" href="#MultiVCF"></a>
</h4>
<p>This mode requires two pieces of information:</p>
<ul>
<li>The path to the folder that contains one or more VCF file(s).</li>
<li>The naming pattern of VCF file(s), with the following requirement:
<ul>
<li>The pattern must include <code>"%s"</code> to declare the
emplacement of the sequence (<em>i.e.</em> chromosome) name in the
pattern.</li>
</ul>
</li>
</ul>
<p>Note that a summary of VCF file(s) detected using the given the
folder and pattern is displayed on the right, to help users determine
whether the parameters are correct. In addition, the content of the
given folder is displayed at the bottom of the page, beside the same
content filtered for the VCF file naming pattern.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<p><em>None</em>. The text fields should already be filled with default
values, pointing to the single example VCF file
(<code>chr15.phase3_integrated.vcf.gz</code>).</p>
</blockquote>
</div>
<div class="section level4">
<h4 id="scanVcfParam">VCF scan parameters<a class="anchor" aria-label="anchor" href="#scanVcfParam"></a>
</h4>
<p>This panel allows users to select the INFO and FORMAT fields to
import (in the <code>info</code> and <code>geno</code> slots of the
<code>VCF</code> object, respectively).</p>
<p>It is important to note that the FORMAT/GT and INFO/<vep>
fields—where <code>&lt;vep&gt;</code> stands for the INFO key where
Ensembl VEP predictions are stored—are implicitely imported from the
VCF. Similarly, the mandatory FIXED fields <code>CHROM</code>,
<code>POS</code>, <code>ID</code>, <code>REF</code>, <code>ALT</code>,
<code>QUAL</code>, and <code>FILTER</code> are automatically imported to
populate the <code>rowRanges</code> slot of the <code>VCF</code>
object.</vep></p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Click the <em>Deselect all</em> action button under the <em>INFO
fields</em> selection input to import only the INFO/CSQ and FORMAT/GT
fields.</li>
<li>Click the <em>Import variants</em> action button</li>
</ul>
<p>A summary of variants, phenotypes, and samples imported will appear
beside the action button.</p>
</blockquote>
</div>
</div>
<div class="section level3">
<h3 id="EnsDbPkg">Annotations<a class="anchor" aria-label="anchor" href="#EnsDbPkg"></a>
</h3>
<p>This panel allows users to select a pre-installed annotation package.
Currently, only <code>EnsDb</code> annotation packages are supported,
and only <strong>gene</strong>-coding regions may be queried.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>If none of the <code>EnsDb</code> packages are installed, it will
simply <strong>not</strong> be possible to use the <code>ensembl</code>
interface of the <em>Genomic ranges</em> input tab.</li>
<li>If the <code>EnsDb.Hsapiens.v75</code> package is the only
<code>EnsDb</code> packages installed, no action is required; the
package should already be pre-selected.</li>
<li>If the <code>EnsDb.Hsapiens.v75</code> package is
<strong>not</strong> the only <code>EnsDb</code> packages installed,
users should select it in the list of choices.</li>
</ul>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="Frequencies">Frequencies panel<a class="anchor" aria-label="anchor" href="#Frequencies"></a>
</h2>
<p>This panel demonstrates the use of three methods implemented in the
<em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
package, namely <code>addFrequencies</code>,
<code>addOverallFrequencies</code>, and
<code>addPhenoLevelFrequencies</code>.</p>
<div class="section level3">
<h3 id="OverallFrequencies">Overall frequencies<a class="anchor" aria-label="anchor" href="#OverallFrequencies"></a>
</h3>
<p>This panel allows users to <em>Add</em> and <em>Remove</em> INFO
fields that contain genotype counts (<em>i.e.</em> homozygote reference,
heterozygote, homozygote alternate) and allele frequencies
(<em>i.e.</em> alternate allele frequency, minor allele frequency)
calculated across all the samples and variants imported. The
web-application uses the homozygote reference, heterozygote, and
homozygote alternate genotypes defined in the <a href="#AdvancedSettings">Advanced settings</a> panel.</p>
<p>Importantly, the name of the INFO keys that are used to store the
calculated values can be defined in the <a href="#AdvancedSettings">Advanced settings</a> panel.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Click the <em>Add</em> action button</li>
<li>See the <em>Latest changes</em> message update at the top of the
screen.</li>
<li>Optionally, the <a href="#Views">Views</a> panel can be used to
examine the new fields</li>
</ul>
</blockquote>
</div>
<div class="section level3">
<h3 id="PhenoLevelFrequencies">Phenotype-level frequencies<a class="anchor" aria-label="anchor" href="#PhenoLevelFrequencies"></a>
</h3>
<p>This panel allows users to <em>Refresh</em> the list of INFO fields
that contain genotype counts and allele frequencies calculated within
<em>groups of samples</em> associated with various levels of a given
phenotype.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Select <code>super_pop</code> in the list of phenotypes</li>
<li>Click the <em>Select all</em> action button</li>
<li>Click the <em>Refresh</em> action button</li>
<li>See the <em>Latest changes</em> message update at the top of the
screen.</li>
<li>Optionally, the <a href="#Views">Views</a> panel can be used to
examine the new fields</li>
</ul>
</blockquote>
</div>
</div>
<div class="section level2">
<h2 id="VcfFilterRules">Filters panel<a class="anchor" aria-label="anchor" href="#VcfFilterRules"></a>
</h2>
<p>One of the flagship features of the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> package
are the <em>VCF filter rules</em>, extending the <em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
<code>FilterRules</code> class to new classes of filter rules that can
be evaluated within environments defined by the various slots of
<code>VCF</code> objects.</p>
<p>Generally speaking, <code>FilterRules</code> greatly facilitate the
design and combination of powerful filter rules for table-like objects,
such as the <code>fixed</code> and <code>info</code> slots of <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
<code>VCF</code> objects, as well as Ensembl VEP predictions stored in
the meta-columns of <code>GRanges</code> returned by the <em><a href="https://bioconductor.org/packages/3.20/ensemblVEP" class="external-link">ensemblVEP</a></em>
<code>parseCSQToGRanges</code> method.</p>
<p>A separate vignette describes in greater detail the use of classes
that contain <em>VCF filter rules</em>. A simple example is shown
below.</p>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>Select <code>VEP</code> as the <em>Type</em> of filter</li>
<li>Paste <code>grepl("missense",Consequence)</code> in the text
field</li>
<li>Leave the <em>Active?</em> checkbox ticked</li>
<li>Click the <em>Add filter</em> action button</li>
<li>See the list of rules update at the bottom of the screen</li>
<li>Click the <em>Apply filters</em> action button</li>
<li>See the summary of filtered variants update beside the action
button</li>
<li>Optionally, the <a href="#Views">Views</a> panel can be used to
examine the new fields</li>
</ul>
<p>Alternatively: click the <em>Sample input</em> button</p>
</blockquote>
</div>
<div class="section level2">
<h2 id="Views">Views panel<a class="anchor" aria-label="anchor" href="#Views"></a>
</h2>
<p>This panel offers the chance to examine the main objects of the
session, namely:</p>
<ul>
<li>The active genomic ranges</li>
<li>The <code>rowRanges</code> and selected meta-columns of the filtered
variants.</li>
<li>Selected field of the <code>info</code> slot (of the filtered
variants).</li>
<li>Selected Ensembl VEP predictions (of the filtered variants).</li>
<li>Selected phenotypes attached to the variants.</li>
<li>Subset of genotypes (among the filtered variants).
<ul>
<li>Genotypes for all filtered variants may be displayed as a heatmap
(<code>ggplot</code>).</li>
</ul>
</li>
</ul>
<blockquote>
<p><strong>Action</strong>:</p>
<ul>
<li>In the various panels, select fields to examine each object
<ul>
<li>In particular, note the INFO fields that contain genotype counts and
allele frequencies calculated <a href="#Frequencies">earlier</a>
</li>
</ul>
</li>
<li>Go to the <em>Heatmap</em> tab of the <em>Genotypes</em> panel</li>
<li>Click the <em>Go!</em> action button to calculate and display the
heatmap</li>
</ul>
</blockquote>
</div>
<div class="section level2">
<h2 id="Plots">Plots panel<a class="anchor" aria-label="anchor" href="#Plots"></a>
</h2>
<p>This panel demonstrates the use of two methods implemented in the
<em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
package, namely <code>tabulateVepByPhenotype</code> and
<code>densityVepByPhenotype</code>.</p>
</div>
<div class="section level2">
<h2 id="settings-panel">Settings panel<a class="anchor" aria-label="anchor" href="#settings-panel"></a>
</h2>
<p>This panel stores more advanced settings that users may not need to
edit as frequently, if at all. Those settings are divided in two
sub-panels:</p>
<ul>
<li>
<strong>Advanced</strong>
<ul>
<li>Genotypes, INFO key suffixes, and VCF yield size</li>
</ul>
</li>
<li>
<strong>Parallel</strong>
<ul>
<li>Use of multiple CPUs to accelerate calculations</li>
</ul>
</li>
</ul>
<div class="section level3">
<h3 id="AdvancedSettings">Advanced settings<a class="anchor" aria-label="anchor" href="#AdvancedSettings"></a>
</h3>
<div class="section level4">
<h4 id="AdvancedGenotypes">Genotypes<a class="anchor" aria-label="anchor" href="#AdvancedGenotypes"></a>
</h4>
<p>It is critical to accurately identify and define how the different
genotypes—homozygote reference, heterozygote, and homozygote
alternate—are encoded in the VCF file, to produce accurate <a href="#Frequencies">genotypes counts and frequencies</a>, for instance.
This generally requires examining the content of the FORMAT/GT field
outside of the web-application. For instance, the functions
<code>unique</code> and <code>table</code> may be used to identify (and
count) all the distinct genotype codes in the <code>geno</code> slot
(<code>"GT"</code> key) of a <code>VCF</code> object.</p>
<p>The default selected values are immediately compatible with the
demonstration data set. Users who wish to select genotypes codes not yet
available among the current choices may either contact the package
maintainer to add them in a future release, or edit the <a href="#GlobalConfig">Global configuration file</a> of the
web-application locally.</p>
</div>
<div class="section level4">
<h4 id="AdvancedSuffixes">INFO key suffixes<a class="anchor" aria-label="anchor" href="#AdvancedSuffixes"></a>
</h4>
<p>Currently, the three calculated genotypes counts and two allele
frequencies require five INFO fields to store their respective
values.</p>
<p>Considering that <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> offers
the possibility to calculate counts and frequencies for the overall data
set, and for each level of each phenotype, it is important to define a
clear and consistent naming mechanism that does not conflict with INFO
keys imported from the VCF file(s). In the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
package, a suffix is required for each type of genotype and frequency
calculated, to generate INFO as follows:</p>
<ul>
<li>Overall counts and frequencies are stored in INFO keys named
<code>&lt;suffix&gt;</code>
</li>
<li>Counts and frequencies calculated for individual levels of selected
phenotypes are stored under INFO keys formed as
<code>&lt;phenotype&gt;_&lt;level&gt;_&lt;suffix&gt;</code>
</li>
</ul>
<p>Again, the default values are immediately compatible with the
demonstration data set. For other data sets, it may be necessary to
change those values, either by preference, or to avoid conflict with
INFO keys imported from the VCF file(s).</p>
</div>
<div class="section level4">
<h4 id="AdvancedMiscellaneous">Miscellaneous settings<a class="anchor" aria-label="anchor" href="#AdvancedMiscellaneous"></a>
</h4>
<p>Other rarely used settings in this panel include:</p>
<ul>
<li>VCF yield size
<ul>
<li>Only applicable when VCF file(s) are parsed without defined <a href="#InputGRanges">genomic ranges</a>. See the <em><a href="https://bioconductor.org/packages/3.20/Rsamtools" class="external-link">Rsamtools</a></em>
documentation.</li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section level3">
<h3 id="ParallelSettings">Parallel settings<a class="anchor" aria-label="anchor" href="#ParallelSettings"></a>
</h3>
<p>Several functionalities of the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> package
are applied to independent subsets of data (<em>e.g.</em> counting
genotypes in various levels of a given phenotype). Such processes can
benefit from multi-threaded calculations. Multi-threading settings in
the Shiny web-application are somewhat experimental, as they have been
validated only on a small set of operating systems, while some issues
have been reported for others.</p>
<table class="table">
<thead><tr class="header">
<th align="center">Report</th>
<th align="center">Operating System</th>
<th align="center">Cluster Class</th>
<th align="center">Cluster type</th>
<th align="center"># Cores</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center">OK</td>
<td align="center">Ubuntu 14.04</td>
<td align="center">Multicore</td>
<td align="center">FORK</td>
<td align="center">2</td>
</tr>
<tr class="even">
<td align="center">OK</td>
<td align="center">Scientific Linux 6.7</td>
<td align="center">Multicore</td>
<td align="center">FORK</td>
<td align="center">2</td>
</tr>
<tr class="odd">
<td align="center">Hang<sub>1</sub>
</td>
<td align="center">OS X El Capitan</td>
<td align="center">Snow</td>
<td align="center">SOCK</td>
<td align="center">2</td>
</tr>
</tbody>
</table>
<ol style="list-style-type: decimal">
<li>Application hangs while CPUs work infinitely at full capacity.</li>
</ol>
</div>
</div>
<div class="section level2">
<h2 id="SessionInfo">Session information<a class="anchor" aria-label="anchor" href="#SessionInfo"></a>
</h2>
<p>The last panel of the Shiny Variant Explorer offers detailed views of
objects and settings in the current session, including:</p>
<ul>
<li>
<strong>Session info</strong>
<ul>
<li>The <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> value</li>
</ul>
</li>
<li>
<strong>TVTB settings</strong>
<ul>
<li>See the vignette <em>Introduction to TVTB</em> for more
information</li>
</ul>
</li>
<li>
<strong>General settings</strong>
<ul>
<li>Including the current value of various input widgets</li>
</ul>
</li>
<li>
<strong>Advanced settings</strong>
<ul>
<li>including the current value of more input widgets</li>
</ul>
</li>
<li>
<strong>BED</strong>
<ul>
<li>Structural view of the active genomic ranges</li>
</ul>
</li>
<li>
<strong>Variants</strong>
<ul>
<li>Overview of the raw <code>VCF</code> object</li>
</ul>
</li>
<li>
<strong>VEP</strong>
<ul>
<li>Structural view of the <code>GRanges</code> that store the Ensembl
VEP predictions</li>
</ul>
</li>
<li>
<strong>Phenotypes</strong>
<ul>
<li>Structural view of the phenotype information attached to the
variants</li>
</ul>
</li>
<li>
<strong>Genotypes</strong>
<ul>
<li>Structural view of the <code>geno</code> slot (<code>"GT"</code>
key) of the raw variants</li>
</ul>
</li>
</ul>
</div>
<div class="section level2">
<h2 id="GlobalConfig">Global configuration<a class="anchor" aria-label="anchor" href="#GlobalConfig"></a>
</h2>
<p>Most default values are stored in the <code>global.R</code> file of
the web-application. All the files of the web-application are stored in
the <code>extdata/shinyApp</code> folder of the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
installation directory (see an <a href="#InputPhenotypes">earlier
section</a> to identify this directory).</p>
<p>Users who wish to change the default values of certain input widgets
(<em>e.g.</em> genotype codes) may edit the <code>global.R</code> file
accordingly. However, the file will be reset at each package update.</p>
</div>
<div class="section level2">
<h2 id="VignetteSessionInfo">Vignette session<a class="anchor" aria-label="anchor" href="#VignetteSessionInfo"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo</a></span><span class="op">(</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] BiocStyle_2.33.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##  [1] digest_0.6.36       desc_1.4.3          R6_2.5.1           </span></span>
<span><span class="co">##  [4] bookdown_0.40       fastmap_1.2.0       xfun_0.45          </span></span>
<span><span class="co">##  [7] cachem_1.1.0        knitr_1.47          htmltools_0.5.8.1  </span></span>
<span><span class="co">## [10] rmarkdown_2.27      lifecycle_1.0.4     cli_3.6.3          </span></span>
<span><span class="co">## [13] sass_0.4.9          pkgdown_2.0.9.9000  textshaping_0.4.0  </span></span>
<span><span class="co">## [16] jquerylib_0.1.4     systemfonts_1.1.0   compiler_4.4.1     </span></span>
<span><span class="co">## [19] tools_4.4.1         ragg_1.3.2          bslib_0.7.0        </span></span>
<span><span class="co">## [22] evaluate_0.24.0     yaml_2.3.8          BiocManager_1.30.23</span></span>
<span><span class="co">## [25] jsonlite_1.8.8      rlang_1.1.4         fs_1.6.4           </span></span>
<span><span class="co">## [28] htmlwidgets_1.6.4</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="References">References<a class="anchor" aria-label="anchor" href="#References"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-RN1" class="csl-entry">
McLaren, W., B. Pritchard, D. Rios, Y. Chen, P. Flicek, and F.
Cunningham. 2010. <span>“Deriving the Consequences of Genomic Variants
with the Ensembl <span>API</span> and <span>SNP</span> Effect
Predictor.”</span> Journal Article. <em>Bioinformatics</em> 26 (16):
2069–70. <a href="https://doi.org/10.1093/bioinformatics/btq330" class="external-link">https://doi.org/10.1093/bioinformatics/btq330</a>.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>The <code>...</code> argument is called “ellipsis”.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>In the future, the web-application may also support
<code>TxDb</code> and <code>OrganismDb</code> annotation packages.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
<li id="fn3"><p>In the future, the web-application may also use
annotation packages to facet statistics and figures by genomic
range(s).<a href="#fnref3" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
