<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>VCF filter rules • TVTB</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="VCF filter rules">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TVTB</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">1.31.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/Introduction.html">Introduction to TVTB</a>
    </li>
    <li>
      <a href="../articles/tSVE.html">The Shiny Variant Explorer</a>
    </li>
    <li>
      <a href="../articles/VcfFilterRules.html">VCF filter rules</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/kevinrue/TVTB/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>VCF filter rules</h1>
                        <h4 data-toc-skip class="author">Kévin
Rue-Albrecht</h4>
            <address class="author_afil">
      Department of Medicine, Imperial College London, UKNuffield
Department of Medicine, University of Oxford,
UK<br><a class="author_email" href="mailto:#"></a><a href="mailto:kevinrue67@gmail.com" class="email">kevinrue67@gmail.com</a>
      </address>
                  
            <h4 data-toc-skip class="date">4 July 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/kevinrue/TVTB/blob/devel/vignettes/VcfFilterRules.Rmd" class="external-link"><code>vignettes/VcfFilterRules.Rmd</code></a></small>
      <div class="hidden name"><code>VcfFilterRules.Rmd</code></div>

    </div>

    
        <div class="abstract">
      <p class="abstract">Abstract</p>
      <p>Currently, <code>VCF</code> objects of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
      package may be subsetted by indices or names of variant records.
      The <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em>
      package extends <code>FilterRules</code> of the <em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
      package to provide filter rules readily applicable to individual
      slots of <code>VCF</code> objects. These new classes of filter
      rules provide containers for powerful expressions and functions to
      facilitate filtering of variants using combinations of filters
      applicable to FIXED fields, INFO fields, and Ensembl VEP
      predictions stored in a given INFO field of VCF files.</p>
    </div>
    
<div class="section level2">
<h2 id="Motivation">Motivation<a class="anchor" aria-label="anchor" href="#Motivation"></a>
</h2>
<div class="section level3">
<h3 id="background">Background<a class="anchor" aria-label="anchor" href="#background"></a>
</h3>
<p><code>VCF</code> objects of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
package contain a plethora of information imported from specific fields
of source VCF files and stored in dedicated slots (<em>e.g.</em>
<code>fixed</code>, <code>info</code>, <code>geno</code>), as well as
optional Ensembl VEP predictions <span class="citation">(McLaren et al.
2010)</span> stored under a given key of their INFO slot.</p>
<p>This information may be used to identify and filter variants of
interest for further analysis. However, the size of genetic data sets
and the variety of filter rules—and their combinatorial explosion—create
considerable challenges in terms of workspace memory and entropy
(<em>i.e.</em> size and number of objects in the workspace,
respectively).</p>
<p>The <code>FilterRules</code> class implemented in the <em><a href="https://bioconductor.org/packages/3.20/S4Vectors" class="external-link">S4Vectors</a></em>
package provides a powerful tool to create flexible and lightweight
filter rules defined in the form of <code>expression</code> and
<code>function</code> objects that can be evaluated within given
environments. The <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> package
extends this <code>FilterRules</code> class into novel classes of
<em>VCF filter rules</em>, applicable to information stored in the
distinct slots of <code>VCF</code> objects (<em>i.e.</em>
<code>CollapsedVCF</code> and <code>ExpandedVCF</code> classes), as
described below:</p>
<table style="width:93%;" class="table">
<caption>Motivation for each of the new classes extending
<code>FilterRules</code>, to define VCF filter rules.</caption>
<colgroup>
<col width="26%">
<col width="66%">
</colgroup>
<thead><tr class="header">
<th align="center">Class</th>
<th align="center">Motivation</th>
</tr></thead>
<tbody>
<tr class="odd">
<td align="center"><code>VcfFixedRules</code></td>
<td align="center">Filter rules applied to the <code>fixed</code> slot
of a <code>VCF</code> object.</td>
</tr>
<tr class="even">
<td align="center"><code>VcfInfoRules</code></td>
<td align="center">Filter rules applied to the <code>info</code> slot of
a <code>VCF</code> object.</td>
</tr>
<tr class="odd">
<td align="center"><code>VcfVepRules</code></td>
<td align="center">Filter rules applied to the Ensembl VEP predictions
stored in a given INFO key of a <code>VCF</code> object.</td>
</tr>
<tr class="even">
<td align="center"><code>VcfFilterRules</code></td>
<td align="center">Combination of <code>VcfFixedRules</code>,
<code>VcfInfoRules</code>, and <code>VcfVepRules</code> applicable to a
<code>VCF</code> object.</td>
</tr>
</tbody>
</table>
<p>Table: Motivation for each of the new classes extending
<code>FilterRules</code> to define <em>VCF filter rules</em>.</p>
<p>Note that <code>FilterRules</code> objects themselves are applicable
to <code>VCF</code> objects, with two important difference from the
above specialised classes:</p>
<ul>
<li>Expressions must explicitely refer to the different <code>VCF</code>
slots</li>
<li>As a consequence, a single expression can refer to fields from
different <code>VCF</code> slots, for instance:</li>
</ul>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fr</span> <span class="op">&lt;-</span> <span class="fu">S4Vectors</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/S4Vectors/man/FilterRules-class.html" class="external-link">FilterRules</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    mixed <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span></span>
<span>        <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">fixed</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span>,<span class="st">"FILTER"</span><span class="op">]</span> <span class="op">==</span> <span class="st">"PASS"</span> <span class="op">&amp;</span></span>
<span>            <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">[</span>,<span class="st">"MAF"</span><span class="op">]</span> <span class="op">&gt;=</span> <span class="fl">0.05</span></span>
<span>    <span class="op">}</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fr</span></span></code></pre></div>
<pre><code><span><span class="co">## FilterRules of length 1</span></span>
<span><span class="co">## names(1): mixed</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="Features">Features<a class="anchor" aria-label="anchor" href="#Features"></a>
</h3>
<p>As they inherit from the <code>FilterRules</code> class, these new
classes benefit from accessors and methods defined for their parent
class, including:</p>
<ul>
<li>
<em>VCF filter rules</em> can be toggled individually between an
active and an inactive states</li>
<li>
<em>VCF filter rules</em> can be subsetted, edited, replaced, and
deleted</li>
</ul>
<p>To account for the more complex structure of <code>VCF</code>
objects, some of the new <em>VCF filter rules</em> classes implemented
in the <em><a href="https://bioconductor.org/packages/3.20/TVTB" class="external-link">TVTB</a></em> package
require additional information stored in new dedicated slots, associated
with the appropriate accessors and setters. For instance:</p>
<ul>
<li>
<code>VcfVepRules</code> require the INFO key where predictions of
the Ensembl Variant Effect Predictor are stored in a <code>VCF</code>
object. The <code>vep</code> accessor method may be used to access this
slot.</li>
<li>
<code>VcfFilterRules</code>—which may combine any number of filter
rules stored in <code>FixedRules</code>, <code>VcfFixedRules</code>,
<code>VcfInfoRules</code>, <code>VcfVepRules</code>, and other
<code>VcfFilterRules</code> objects— mark each filter rule with their
type in the combined object. The information is stored in the
<code>type</code> slot, which may be accessed using the
<em>read-only</em> accessor method <code>type</code>.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="DemoData">Demonstration data<a class="anchor" aria-label="anchor" href="#DemoData"></a>
</h2>
<p>For the purpose of demonstrating the utility and usage of <em>VCF
filter rules</em>, a set of variants and associated phenotype
information was obtained from the <a href="http://www.1000genomes.org" class="external-link">1000 Genomes Project</a> Phase 3
release. It can be imported as a <code>CollapsedVCF</code> object using
the following code:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/kevinrue/TVTB" class="external-link">TVTB</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Registered S3 method overwritten by 'GGally':</span></span>
<span><span class="co">##   method from   </span></span>
<span><span class="co">##   +.gg   ggplot2</span></span></code></pre>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">extdata</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, package <span class="op">=</span> <span class="st">"TVTB"</span><span class="op">)</span></span>
<span><span class="va">vcfFile</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">extdata</span>, <span class="st">"chr15.phase3_integrated.vcf.gz"</span><span class="op">)</span></span>
<span><span class="va">tabixVcf</span> <span class="op">&lt;-</span> <span class="fu">Rsamtools</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/Rsamtools/man/TabixFile-class.html" class="external-link">TabixFile</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">vcfFile</span><span class="op">)</span></span>
<span><span class="va">vcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/readVcf-methods.html" class="external-link">readVcf</a></span><span class="op">(</span>file <span class="op">=</span> <span class="va">tabixVcf</span><span class="op">)</span></span></code></pre></div>
<p>VCF filter rules may be applied to <code>ExpandedVCF</code> objects
equally:</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">evcf</span> <span class="op">&lt;-</span> <span class="fu">VariantAnnotation</span><span class="fu">::</span><span class="fu">expand</span><span class="op">(</span>x <span class="op">=</span> <span class="va">vcf</span>, row.names <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="section level3">
<h3 id="VcfClasses">CollapsedVCF and ExpandedVCF<a class="anchor" aria-label="anchor" href="#VcfClasses"></a>
</h3>
<p>As described in the documentation of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
package, the key difference between <code>CollapsedVCF</code> and
<code>ExpandedVCF</code> objects —both extending the <code>VCF</code>
class—is the expansion of multi-allelic records into bi-allelic records,
respectively. In other words (quoting the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
documentation):</p>
<blockquote>
<p>“<code>CollapsedVCF</code> objects contains the ALT data as a
<code>DNAStringSetList</code> allowing for multiple alleles per variant.
In contrast, the <code>ExpandedVCF</code> stores the ALT data as a
<code>DNAStringSet</code> where the ALT column has been expanded to
create a flat form of the data with one row per variant-allele
combination.”</p>
</blockquote>
<p>This difference has implications for filter rules using the
<code>"ALT"</code> field of the <code>info</code> slot, as demonstrated
in a <a href="#AltField">later section</a>.</p>
</div>
</div>
<div class="section level2">
<h2 id="VcfFields">Fields available for the definition of filter rules<a class="anchor" aria-label="anchor" href="#VcfFields"></a>
</h2>
<p>First, let us examine which fields (<em>i.e.</em> column names) are
available in the <code>VCF</code> objects to create <em>VCF filter
rules</em>:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fixedVcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">fixed</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fixedVcf</span></span></code></pre></div>
<pre><code><span><span class="co">## [1] "REF"    "ALT"    "QUAL"   "FILTER"</span></span></code></pre>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">infoVcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">infoVcf</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "CIEND"         "CIPOS"         "CS"            "END"          </span></span>
<span><span class="co">##  [5] "IMPRECISE"     "MC"            "MEINFO"        "MEND"         </span></span>
<span><span class="co">##  [9] "MLEN"          "MSTART"        "SVLEN"         "SVTYPE"       </span></span>
<span><span class="co">## [13] "TSD"           "AC"            "AF"            "NS"           </span></span>
<span><span class="co">## [17] "AN"            "EAS_AF"        "EUR_AF"        "AFR_AF"       </span></span>
<span><span class="co">## [21] "AMR_AF"        "SAS_AF"        "DP"            "AA"           </span></span>
<span><span class="co">## [25] "VT"            "EX_TARGET"     "MULTI_ALLELIC" "CSQ"</span></span></code></pre>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">csq</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/parseCSQToGRanges.html">parseCSQToGRanges</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">evcf</span><span class="op">)</span></span>
<span><span class="va">vepVcf</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="fu">mcols</span><span class="op">(</span><span class="va">csq</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">vepVcf</span></span></code></pre></div>
<pre><code><span><span class="co">##  [1] "Allele"             "Consequence"        "IMPACT"            </span></span>
<span><span class="co">##  [4] "SYMBOL"             "Gene"               "Feature_type"      </span></span>
<span><span class="co">##  [7] "Feature"            "BIOTYPE"            "EXON"              </span></span>
<span><span class="co">## [10] "INTRON"             "HGVSc"              "HGVSp"             </span></span>
<span><span class="co">## [13] "cDNA_position"      "CDS_position"       "Protein_position"  </span></span>
<span><span class="co">## [16] "Amino_acids"        "Codons"             "Existing_variation"</span></span>
<span><span class="co">## [19] "DISTANCE"           "STRAND"             "FLAGS"             </span></span>
<span><span class="co">## [22] "VARIANT_CLASS"      "SYMBOL_SOURCE"      "HGNC_ID"           </span></span>
<span><span class="co">## [25] "CANONICAL"          "TSL"                "APPRIS"            </span></span>
<span><span class="co">## [28] "CCDS"               "ENSP"               "SWISSPROT"         </span></span>
<span><span class="co">## [31] "TREMBL"             "UNIPARC"            "GENE_PHENO"        </span></span>
<span><span class="co">## [34] "SIFT"               "PolyPhen"           "DOMAINS"           </span></span>
<span><span class="co">## [37] "HGVS_OFFSET"        "GMAF"               "AFR_MAF"           </span></span>
<span><span class="co">## [40] "AMR_MAF"            "EAS_MAF"            "EUR_MAF"           </span></span>
<span><span class="co">## [43] "SAS_MAF"            "AA_MAF"             "EA_MAF"            </span></span>
<span><span class="co">## [46] "ExAC_MAF"           "ExAC_Adj_MAF"       "ExAC_AFR_MAF"      </span></span>
<span><span class="co">## [49] "ExAC_AMR_MAF"       "ExAC_EAS_MAF"       "ExAC_FIN_MAF"      </span></span>
<span><span class="co">## [52] "ExAC_NFE_MAF"       "ExAC_OTH_MAF"       "ExAC_SAS_MAF"      </span></span>
<span><span class="co">## [55] "CLIN_SIG"           "SOMATIC"            "PHENO"             </span></span>
<span><span class="co">## [58] "PUBMED"             "MOTIF_NAME"         "MOTIF_POS"         </span></span>
<span><span class="co">## [61] "HIGH_INF_POS"       "MOTIF_SCORE_CHANGE" "CADD_PHRED"        </span></span>
<span><span class="co">## [64] "CADD_RAW"</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="FilterUsage">Usage of VCF filter rules<a class="anchor" aria-label="anchor" href="#FilterUsage"></a>
</h2>
<div class="section level3">
<h3 id="SingleFieldFilter">Filter rules using a single field<a class="anchor" aria-label="anchor" href="#SingleFieldFilter"></a>
</h3>
<p>The value of a particular field can be used to define expressions
that represent simple filter rules based on that value alone. Multiple
rules may be stored in any one <code>FilterRules</code> objects.
Ideally, <em>VCF filter rules</em> should be named to facilitate their
use, but also as a reminder of the purpose of each particular rule. For
instance, in the chunk of code below, two filter rules are defined using
fields of the <code>fixed</code> slot:</p>
<ul>
<li>A rule named <code>"pass"</code> identifies variants for which the
value in the <code>FILTER</code> field is <code>"PASS"</code>
</li>
<li>A rule named <code>"qual20"</code> identifies variants where the
value in the <code>QUAL</code> field is greater than or equal to
<code>20</code>
</li>
</ul>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fixedRules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfFixedRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    pass <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">FILTER</span> <span class="op">==</span> <span class="st">"PASS"</span><span class="op">)</span>,</span>
<span>    qual20 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">QUAL</span> <span class="op">&gt;=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu">active</span><span class="op">(</span><span class="va">fixedRules</span><span class="op">)</span><span class="op">[</span><span class="st">"qual20"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">fixedRules</span>, <span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    pass          qual20       </span></span>
<span><span class="co">##  Mode:logical   Mode:logical  </span></span>
<span><span class="co">##  TRUE:479       TRUE:479</span></span></code></pre>
<p>In the example above, all variants pass the active
<code>"pass"</code> filter, while the deactivated rules
<code>"qual20"</code>) automatically returns <code>TRUE</code> for all
variants.</p>
</div>
<div class="section level3">
<h3 id="MultipleFieldsFilter">Filter rules using multiple fields<a class="anchor" aria-label="anchor" href="#MultipleFieldsFilter"></a>
</h3>
<p>It is also possible for <em>VCF filter rules</em> to use multiple
fields (of the same <code>VCF</code> slot) in a single expression. In
the chunk of code below, the <em>VCF filter rule</em> identifies
variants for which both the <code>"REF"</code> and <code>"ALT"</code>
values (in the INFO slot) are one of the four nucleotides (<em>i.e.</em>
a simple definition of single nucleotide polymorphisms; SNPs):</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">nucleotides</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"A"</span>, <span class="st">"T"</span>, <span class="st">"G"</span>, <span class="st">"C"</span><span class="op">)</span></span>
<span><span class="va">SNPrule</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfFixedRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    SNP <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">REF</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">nucleotides</span> <span class="op">&amp;</span></span>
<span>        <span class="fu"><a href="https://rdrr.io/r/base/character.html" class="external-link">as.character</a></span><span class="op">(</span><span class="va">ALT</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="va">nucleotides</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">SNPrule</span>, <span class="va">evcf</span>, enclos <span class="op">=</span> <span class="va">.GlobalEnv</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##     SNP         </span></span>
<span><span class="co">##  Mode :logical  </span></span>
<span><span class="co">##  FALSE:14       </span></span>
<span><span class="co">##  TRUE :467</span></span></code></pre>
<p>Some considerations regarding the above filter rule:</p>
<ul>
<li>considering that the filter rule requires the
<code>nucleotides</code> character vector, the global environment must
be supplied as the enclosing environment to successfully evaluate the
expression</li>
<li>
<code>"REF"</code> and <code>"ALT"</code> are stored as
<code>DNAStringSet</code> in <code>CollapsedVCF</code> objects and must
be converted to <code>character</code> in order to successfully apply
the method <code>%in%</code>.</li>
</ul>
</div>
<div class="section level3">
<h3 id="CalculationFilter">Calculations in filter rules<a class="anchor" aria-label="anchor" href="#CalculationFilter"></a>
</h3>
<p>Expressions that define filter rules may also include calculations.
In the chunk of code below, two simple <em>VCF filter rules</em> are
defined using fields of the <code>info</code> slot:</p>
<ul>
<li>A rule named <code>"samples"</code> identifies variants where at
least 90% of samples have data (<em>i.e.</em> the <code>NS</code> value
is greater than or equal to <code>0.9</code> times the total number of
samples)</li>
<li>A rule named <code>"avgSuperPopAF"</code> calculates the average of
the allele frequencies calculated in each the five super-populations
(available in several INFO fields), and subsequently identifies variants
with an average value greater than <code>0.05</code>.</li>
</ul>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">infoRules</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfInfoRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    samples <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">NS</span> <span class="op">&gt;</span> <span class="op">(</span><span class="fl">0.9</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">evcf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>    avgSuperPopAF <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span></span>
<span>        <span class="op">(</span><span class="va">EAS_AF</span> <span class="op">+</span> <span class="va">EUR_AF</span> <span class="op">+</span> <span class="va">AFR_AF</span> <span class="op">+</span> <span class="va">AMR_AF</span> <span class="op">+</span> <span class="va">SAS_AF</span><span class="op">)</span> <span class="op">/</span> <span class="fl">5</span> <span class="op">&gt;</span> <span class="fl">0.05</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">infoRules</span>, <span class="va">evcf</span>, enclos <span class="op">=</span> <span class="va">.GlobalEnv</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  samples        avgSuperPopAF  </span></span>
<span><span class="co">##  Mode:logical   Mode :logical  </span></span>
<span><span class="co">##  TRUE:481       FALSE:452      </span></span>
<span><span class="co">##                 TRUE :29</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="FunctionFilter">Functions in filter rules<a class="anchor" aria-label="anchor" href="#FunctionFilter"></a>
</h3>
<p>It may be more convenient to define filters as <code>function</code>
objects. For instance, the chunk of code below:</p>
<ul>
<li>first, defines a function that:
<ul>
<li>expects the <code>info</code> slot of a <code>VCF</code> object as
input</li>
<li>identifies variants where at least two thirds of the
super-populations show an allele frequency greater than 5%</li>
</ul>
</li>
<li>next, defines a <em>VCF filter rule</em> using the above
function</li>
</ul>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">AFcutoff</span> <span class="op">&lt;-</span> <span class="fl">0.05</span></span>
<span><span class="va">popCutoff</span> <span class="op">&lt;-</span> <span class="fl">2</span><span class="op">/</span><span class="fl">3</span></span>
<span><span class="va">filterFUN</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">envir</span><span class="op">)</span><span class="op">{</span></span>
<span>    <span class="co"># info(vcf) returns a DataFrame; rowSums below requires a data.frame</span></span>
<span>    <span class="va">df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">envir</span><span class="op">)</span></span>
<span>    <span class="co"># Identify fields storing allele frequency in super-populations</span></span>
<span>    <span class="va">popFreqCols</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grep</a></span><span class="op">(</span><span class="st">"[[:alpha:]]{3}_AF"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">df</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># Count how many super-population have an allele freq above the cutoff</span></span>
<span>    <span class="va">popCount</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/MatrixGenerics/man/rowSums.html" class="external-link">rowSums</a></span><span class="op">(</span><span class="va">df</span><span class="op">[</span>,<span class="va">popFreqCols</span><span class="op">]</span> <span class="op">&gt;</span> <span class="va">AFcutoff</span><span class="op">)</span></span>
<span>    <span class="co"># Convert the cutoff ratio to a integer count</span></span>
<span>    <span class="va">popCutOff</span> <span class="op">&lt;-</span> <span class="va">popCutoff</span> <span class="op">*</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">popFreqCols</span><span class="op">)</span></span>
<span>    <span class="co"># Identifies variants where enough super-population pass the cutoff</span></span>
<span>    <span class="va">testRes</span> <span class="op">&lt;-</span> <span class="op">(</span><span class="va">popCount</span> <span class="op">&gt;</span> <span class="va">popCutOff</span><span class="op">)</span></span>
<span>    <span class="co"># Return a boolean vector, required by the eval method</span></span>
<span>    <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">testRes</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">funFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfInfoRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    commonSuperPops <span class="op">=</span> <span class="va">filterFUN</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">funFilter</span>, <span class="va">evcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  commonSuperPops</span></span>
<span><span class="co">##  Mode :logical  </span></span>
<span><span class="co">##  FALSE:464      </span></span>
<span><span class="co">##  TRUE :17</span></span></code></pre>
<p>Notably, the <code>filterFUN</code> function may also be applied
separately to the <code>info</code> slot of <code>VCF</code>
objects:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">filterFUN</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">info</a></span><span class="op">(</span><span class="va">evcf</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE    TRUE </span></span>
<span><span class="co">## logical     464      17</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="PatternFilter">Pattern matching in filter rules<a class="anchor" aria-label="anchor" href="#PatternFilter"></a>
</h3>
<p>The <code>grepl</code> function is particularly suited for the
purpose of <code>FilterRules</code> as they return a
<code>logical</code> vector:</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">missenseFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfVepRules</a></span><span class="op">(</span></span>
<span>    exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>        exact <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">Consequence</span> <span class="op">==</span> <span class="st">"missense_variant"</span><span class="op">)</span>,</span>
<span>        grepl <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/grep.html" class="external-link">grepl</a></span><span class="op">(</span><span class="st">"missense"</span>, <span class="va">Consequence</span><span class="op">)</span><span class="op">)</span></span>
<span>        <span class="op">)</span>,</span>
<span>    vep <span class="op">=</span> <span class="st">"CSQ"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">missenseFilter</span>, <span class="va">evcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    exact           grepl        </span></span>
<span><span class="co">##  Mode :logical   Mode :logical  </span></span>
<span><span class="co">##  FALSE:454       FALSE:452      </span></span>
<span><span class="co">##  TRUE :27        TRUE :29</span></span></code></pre>
<p>In the above chunk of code:</p>
<ul>
<li>the filter rule named <code>"exact"</code> matches only the given
value, associated with <code>27</code> variants,</li>
<li>the filter rule named <code>"grepl"</code> also matches an extra two
variants associated with the value
<code>"missense_variant&amp;splice_region_variant"</code> matched by the
given pattern. By deduction, the two rules indicate together that those
two variants were not assigned the <code>"missense_variant"</code>
prediction.</li>
</ul>
</div>
</div>
<div class="section level2">
<h2 id="AltField">Using ALT data in the <code>fixed</code> slot of <code>VCF</code>
objects<a class="anchor" aria-label="anchor" href="#AltField"></a>
</h2>
<p>As detailed in an earlier section introducing the <a href="#VcfClasses">demonstration data</a>, and more thoroughly in the
documentation of the <em><a href="https://bioconductor.org/packages/3.20/VariantAnnotation" class="external-link">VariantAnnotation</a></em>
package, <code>CollapsedVCF</code> and <code>ExpandedVCF</code> classes
differ in the class of data stored in the <code>"ALT"</code> field of
their respective <code>fixed</code> slot. As as result, <em>VCF filter
rules</em> using data from this field must take into account the
<code>VCF</code> class in order to handle the data appropriately:</p>
<div class="section level3">
<h3 id="ALTExpandedVCF">
<code>ExpandedVCF</code> objects<a class="anchor" aria-label="anchor" href="#ALTExpandedVCF"></a>
</h3>
<p>A key aspect of <code>ExpandedVCF</code> objects is that the
<code>"ALT"</code> field of their <code>fixed</code> slot may store only
a single allele per record as a <code>DNAStringSet</code> object.</p>
<p>For instance, in an earlier section that demonstrated <a href="#MultipleFieldsFilter">Filter rules using multiple raw fields</a>,
ALT data of the <code>fixed</code> slot in an <code>ExpandedVCF</code>
object had to be re-typed from <code>DNAStringSet</code> to
<code>character</code> before the <code>%in%</code> function could be
applied.</p>
<p>Nevertheless, <em>VCF filter rules</em> may also make use of methods
associated with the <code>DNAStringSet</code> class. For instance,
genetic insertions may be identified using the fields <code>"REF"</code>
and <code>"ALT"</code> fields of the <code>fixed</code> slot:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fixedInsertionFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfFixedRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    isInsertion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span></span>
<span>        <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu">width</span><span class="op">(</span><span class="va">ALT</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fu">Biostrings</span><span class="fu">::</span><span class="fu">width</span><span class="op">(</span><span class="va">REF</span><span class="op">)</span></span>
<span>    <span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">evcf_fixedIns</span> <span class="op">&lt;-</span> <span class="fu">subsetByFilter</span><span class="op">(</span><span class="va">evcf</span>, <span class="va">fixedInsertionFilter</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">fixed</a></span><span class="op">(</span><span class="va">evcf_fixedIns</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"REF"</span>, <span class="st">"ALT"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   REF ALT</span></span>
<span><span class="co">## 1   A  AC</span></span>
<span><span class="co">## 2   A  AT</span></span>
<span><span class="co">## 3   C  CA</span></span>
<span><span class="co">## 4   T  TA</span></span></code></pre>
<p>Here, the above <code>VcfFixedRules</code> is synonym to a distinct
<code>VcfVepRules</code> using the Ensembl VEP prediction
<code>"VARIANT_CLASS"</code>:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vepInsertionFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfVepRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    isInsertion <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="va">VARIANT_CLASS</span> <span class="op">==</span> <span class="st">"insertion"</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">evcf_vepIns</span> <span class="op">&lt;-</span> <span class="fu">subsetByFilter</span><span class="op">(</span><span class="va">evcf</span>, <span class="va">vepInsertionFilter</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/VariantAnnotation/man/VCF-class.html" class="external-link">fixed</a></span><span class="op">(</span><span class="va">evcf_vepIns</span><span class="op">)</span><span class="op">[</span>,<span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"REF"</span>, <span class="st">"ALT"</span><span class="op">)</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##   REF ALT</span></span>
<span><span class="co">## 1   A  AC</span></span>
<span><span class="co">## 2   A  AT</span></span>
<span><span class="co">## 3   C  CA</span></span>
<span><span class="co">## 4   T  TA</span></span></code></pre>
</div>
<div class="section level3">
<h3 id="ALTCollapsedVCF">
<code>CollapsedVCF</code> objects<a class="anchor" aria-label="anchor" href="#ALTCollapsedVCF"></a>
</h3>
<p>In contrast to <code>ExpandedVCF</code>, <code>CollapsedVCF</code>
may contain more than one allele per record in their <code>"ALT"</code>
field (<code>fixed</code> slot), represented by a
<code>DNAStringSetList</code> object.</p>
<p>As a result, <em>VCF filter rules</em> using the <code>"ALT"</code>
field of the <code>info</code> slot in <code>CollapsedVCF</code> objects
may use methods dedicated to <code>DNAStringSetList</code> to handle the
data. For instance, multi-allelic variants may be identified by the
following <code>VcfFixedRules</code>:</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">multiallelicFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/TVTB/man/VcfBasicRules-class.html" class="external-link">VcfFixedRules</a></span><span class="op">(</span>exprs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span></span>
<span>    multiallelic <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/expression.html" class="external-link">expression</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/lengths.html" class="external-link">lengths</a></span><span class="op">(</span><span class="va">ALT</span><span class="op">)</span> <span class="op">&gt;</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/eval.html" class="external-link">eval</a></span><span class="op">(</span><span class="va">multiallelicFilter</span>, <span class="va">vcf</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    Mode   FALSE    TRUE </span></span>
<span><span class="co">## logical     477       2</span></span></code></pre>
</div>
</div>
<div class="section level2">
<h2 id="VcfFilterRules">Combination of multiple types of <em>VCF filter rules</em><a class="anchor" aria-label="anchor" href="#VcfFilterRules"></a>
</h2>
<p>Any number of <code>VcfFixedRules</code>, <code>VcfInfoRules</code>,
and <code>VcfVepRules</code>—or even <code>VcfFilterRules</code>
themselves—may be combined into a larger object of class
<code>VcfFilterRules</code>. Notably, the <em>active</em> state of each
filter rule is transferred to the combined object. Even though the
<code>VcfFilterRules</code> class acts as a container for multiple types
of <em>VCF filter rules</em>, the resulting <code>VcfFilterRules</code>
object also extends the <code>FilterRules</code> class, and as a result
can be evaluated and used to subset <code>VCF</code> objects identically
to any of the specialised more specialised classes.</p>
<p>During the creation of <code>VcfFixedRules</code> objects, each
<em>VCF filter rule</em> being combined is marked with a <em>type</em>
value, indicating the VCF slot in which the filter rule must be
evaluated. This information is stored in the new <code>type</code> slot
of <code>VcfFixedRules</code> objects. For instance, it is possible to
combine two <code>VcfFixedRules</code> (containing two and one filter
rules, respectively), one <code>VcfInfoRules</code>, and one
<code>VcfVepRules</code> defined earlier in this vignette:</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">vignetteRules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VcfFilterRules-class.html">VcfFilterRules</a></span><span class="op">(</span></span>
<span>    <span class="va">fixedRules</span>,</span>
<span>    <span class="va">SNPrule</span>,</span>
<span>    <span class="va">infoRules</span>,</span>
<span>    <span class="va">vepInsertionFilter</span></span>
<span><span class="op">)</span></span>
<span><span class="va">vignetteRules</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfFilterRules of length 6</span></span>
<span><span class="co">## names(6): pass qual20 SNP samples avgSuperPopAF isInsertion</span></span></code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">vignetteRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          pass        qual20           SNP       samples avgSuperPopAF </span></span>
<span><span class="co">##          TRUE         FALSE          TRUE          TRUE          TRUE </span></span>
<span><span class="co">##   isInsertion </span></span>
<span><span class="co">##          TRUE</span></span></code></pre>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/type.html" class="external-link">type</a></span><span class="op">(</span><span class="va">vignetteRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          pass        qual20           SNP       samples avgSuperPopAF </span></span>
<span><span class="co">##       "fixed"       "fixed"       "fixed"        "info"        "info" </span></span>
<span><span class="co">##   isInsertion </span></span>
<span><span class="co">##         "vep"</span></span></code></pre>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">vignetteRules</span>, <span class="va">evcf</span>, enclos <span class="op">=</span> <span class="va">.GlobalEnv</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    pass          qual20           SNP          samples        avgSuperPopAF  </span></span>
<span><span class="co">##  Mode:logical   Mode:logical   Mode :logical   Mode:logical   Mode :logical  </span></span>
<span><span class="co">##  TRUE:481       TRUE:481       FALSE:14        TRUE:481       FALSE:452      </span></span>
<span><span class="co">##                                TRUE :467                      TRUE :29       </span></span>
<span><span class="co">##  isInsertion    </span></span>
<span><span class="co">##  Mode :logical  </span></span>
<span><span class="co">##  FALSE:477      </span></span>
<span><span class="co">##  TRUE :4</span></span></code></pre>
<p>Clearly<a href="#fn1" class="footnote-ref" id="fnref1"><sup>1</sup></a>, the <em>VCF filter rules</em>
<code>SNP</code> and <code>isInsertion</code> are mutually exclusive,
which explains the final <code>0</code> variants left after filtering.
Conveniently, either of these rules may be deactivated before evaluating
the remaining active filter rules:</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">vignetteRules</span><span class="op">)</span><span class="op">[</span><span class="st">"SNP"</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">vignetteRules</span>, <span class="va">evcf</span>, enclos <span class="op">=</span> <span class="va">.GlobalEnv</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    pass          qual20          SNP          samples        avgSuperPopAF  </span></span>
<span><span class="co">##  Mode:logical   Mode:logical   Mode:logical   Mode:logical   Mode :logical  </span></span>
<span><span class="co">##  TRUE:481       TRUE:481       TRUE:481       TRUE:481       FALSE:452      </span></span>
<span><span class="co">##                                                              TRUE :29       </span></span>
<span><span class="co">##  isInsertion    </span></span>
<span><span class="co">##  Mode :logical  </span></span>
<span><span class="co">##  FALSE:477      </span></span>
<span><span class="co">##  TRUE :4</span></span></code></pre>
<p>As a result, the deactivated filter rule (<code>"SNP"</code>) now
returns <code>TRUE</code> for all variants, leaving a final
<code>2</code> variants<a href="#fn2" class="footnote-ref" id="fnref2"><sup>2</sup></a> pass the remaining active filter rules:</p>
<ul>
<li>INFO/FILTER equal to <code>"PASS"</code>
</li>
<li>INFO/NS greater than 90% of the number of samples in the data
set</li>
<li>Average of super-population allele frequencies greater than
<code>0.05</code>
</li>
<li>Ensembl VEP prediction <code>VARIANT_CLASS</code> equal to
<code>"insertion"</code>
</li>
</ul>
<p>Finally, the following chunk of code demonstrates how
<code>VcfFilterRules</code> may also be created from the combination of
<code>VcfFilterRules</code>, either with themselves or with any of the
classes that define more specific <em>VCF filter rules</em>. Notably,
when <code>VcfFilterRules</code> objects are combined, the
<code>type</code> and <code>active</code> value of each filter rule is
transferred to the combined object:</p>
<p><strong>Combine <code>VcfFilterRules</code> with
<code>VcfVepRules</code></strong></p>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">combinedFilters</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VcfFilterRules-class.html">VcfFilterRules</a></span><span class="op">(</span></span>
<span>    <span class="va">vignetteRules</span>, <span class="co"># VcfFilterRules</span></span>
<span>    <span class="va">missenseFilter</span> <span class="co"># VcfVepRules</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/type.html" class="external-link">type</a></span><span class="op">(</span><span class="va">vignetteRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          pass        qual20           SNP       samples avgSuperPopAF </span></span>
<span><span class="co">##       "fixed"       "fixed"       "fixed"        "info"        "info" </span></span>
<span><span class="co">##   isInsertion </span></span>
<span><span class="co">##         "vep"</span></span></code></pre>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/type.html" class="external-link">type</a></span><span class="op">(</span><span class="va">combinedFilters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          pass        qual20           SNP       samples avgSuperPopAF </span></span>
<span><span class="co">##       "fixed"       "fixed"       "fixed"        "info"        "info" </span></span>
<span><span class="co">##   isInsertion         exact         grepl </span></span>
<span><span class="co">##         "vep"         "vep"         "vep"</span></span></code></pre>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">vignetteRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          pass        qual20           SNP       samples avgSuperPopAF </span></span>
<span><span class="co">##          TRUE         FALSE         FALSE          TRUE          TRUE </span></span>
<span><span class="co">##   isInsertion </span></span>
<span><span class="co">##          TRUE</span></span></code></pre>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">missenseFilter</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## exact grepl </span></span>
<span><span class="co">##  TRUE  TRUE</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">combinedFilters</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##          pass        qual20           SNP       samples avgSuperPopAF </span></span>
<span><span class="co">##          TRUE         FALSE         FALSE          TRUE          TRUE </span></span>
<span><span class="co">##   isInsertion         exact         grepl </span></span>
<span><span class="co">##          TRUE          TRUE          TRUE</span></span></code></pre>
<p><strong>Combine multiple <code>VcfFilterRules</code> with
<code>VcfFilterRules</code> (and more)</strong></p>
<p>To demonstrate this action, another <code>VcfFilterRules</code> must
first be created. This can be achieve by simply re-typing a
<code>VcfVepRules</code> defined earlier:</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">secondVcfFilter</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VcfFilterRules-class.html">VcfFilterRules</a></span><span class="op">(</span><span class="va">missenseFilter</span><span class="op">)</span></span>
<span><span class="va">secondVcfFilter</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfFilterRules of length 2</span></span>
<span><span class="co">## names(2): exact grepl</span></span></code></pre>
<p>It is now possible to combine the two <code>VcfFilterRules</code>.
Let us even combine another <code>VcfInfoRules</code> object in the same
step:</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">manyRules</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/VcfFilterRules-class.html">VcfFilterRules</a></span><span class="op">(</span></span>
<span>    <span class="va">vignetteRules</span>, <span class="co"># VcfFilterRules</span></span>
<span>    <span class="va">secondVcfFilter</span>, <span class="co"># VcfFilterRules</span></span>
<span>    <span class="va">funFilter</span> <span class="co"># VcfInfoRules</span></span>
<span><span class="op">)</span></span>
<span><span class="va">manyRules</span></span></code></pre></div>
<pre><code><span><span class="co">## VcfFilterRules of length 9</span></span>
<span><span class="co">## names(9): pass qual20 SNP samples avgSuperPopAF isInsertion exact grepl commonSuperPops</span></span></code></pre>
<div class="sourceCode" id="cb55"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">active</span><span class="op">(</span><span class="va">manyRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            pass          qual20             SNP         samples   avgSuperPopAF </span></span>
<span><span class="co">##            TRUE           FALSE           FALSE            TRUE            TRUE </span></span>
<span><span class="co">##     isInsertion           exact           grepl commonSuperPops </span></span>
<span><span class="co">##            TRUE            TRUE            TRUE            TRUE</span></span></code></pre>
<div class="sourceCode" id="cb57"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/type.html" class="external-link">type</a></span><span class="op">(</span><span class="va">manyRules</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            pass          qual20             SNP         samples   avgSuperPopAF </span></span>
<span><span class="co">##         "fixed"         "fixed"         "fixed"          "info"          "info" </span></span>
<span><span class="co">##     isInsertion           exact           grepl commonSuperPops </span></span>
<span><span class="co">##           "vep"           "vep"           "vep"          "info"</span></span></code></pre>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="fu">evalSeparately</span><span class="op">(</span><span class="va">manyRules</span>, <span class="va">evcf</span>, enclos <span class="op">=</span> <span class="va">.GlobalEnv</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##    pass          qual20          SNP          samples        avgSuperPopAF  </span></span>
<span><span class="co">##  Mode:logical   Mode:logical   Mode:logical   Mode:logical   Mode :logical  </span></span>
<span><span class="co">##  TRUE:481       TRUE:481       TRUE:481       TRUE:481       FALSE:452      </span></span>
<span><span class="co">##                                                              TRUE :29       </span></span>
<span><span class="co">##  isInsertion       exact           grepl         commonSuperPops</span></span>
<span><span class="co">##  Mode :logical   Mode :logical   Mode :logical   Mode :logical  </span></span>
<span><span class="co">##  FALSE:477       FALSE:454       FALSE:452       FALSE:464      </span></span>
<span><span class="co">##  TRUE :4         TRUE :27        TRUE :29        TRUE :17</span></span></code></pre>
<p>Critically, users must be careful to combine rules
<strong>all</strong> compatible with the class of <code>VCF</code>
object in which it will be evaluated (<em>i.e.</em>
<code>CollapsedVCF</code> or <code>ExpandedVCF</code>).</p>
</div>
<div class="section level2">
<h2 id="SessionInfo">Session info<a class="anchor" aria-label="anchor" href="#SessionInfo"></a>
</h2>
<p>Here is the output of <code><a href="https://rdrr.io/r/utils/sessionInfo.html" class="external-link">sessionInfo()</a></code> on the system on
which this document was compiled:</p>
<pre><code><span><span class="co">## R version 4.4.1 (2024-06-14)</span></span>
<span><span class="co">## Platform: aarch64-apple-darwin20</span></span>
<span><span class="co">## Running under: macOS Sonoma 14.5</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Matrix products: default</span></span>
<span><span class="co">## BLAS:   /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRblas.0.dylib </span></span>
<span><span class="co">## LAPACK: /Library/Frameworks/R.framework/Versions/4.4-arm64/Resources/lib/libRlapack.dylib;  LAPACK version 3.12.0</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## locale:</span></span>
<span><span class="co">## [1] en_US.UTF-8/en_US.UTF-8/en_US.UTF-8/C/en_US.UTF-8/en_US.UTF-8</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## time zone: UTC</span></span>
<span><span class="co">## tzcode source: internal</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## attached base packages:</span></span>
<span><span class="co">## [1] stats     graphics  grDevices utils     datasets  methods   base     </span></span>
<span><span class="co">## </span></span>
<span><span class="co">## other attached packages:</span></span>
<span><span class="co">## [1] TVTB_1.31.1      BiocStyle_2.33.1</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## loaded via a namespace (and not attached):</span></span>
<span><span class="co">##   [1] RColorBrewer_1.1-3          rstudioapi_0.16.0          </span></span>
<span><span class="co">##   [3] jsonlite_1.8.8              magrittr_2.0.3             </span></span>
<span><span class="co">##   [5] GenomicFeatures_1.57.0      rmarkdown_2.27             </span></span>
<span><span class="co">##   [7] fs_1.6.4                    BiocIO_1.15.0              </span></span>
<span><span class="co">##   [9] zlibbioc_1.51.1             ragg_1.3.2                 </span></span>
<span><span class="co">##  [11] vctrs_0.6.5                 memoise_2.0.1              </span></span>
<span><span class="co">##  [13] Rsamtools_2.21.0            RCurl_1.98-1.14            </span></span>
<span><span class="co">##  [15] base64enc_0.1-3             htmltools_0.5.8.1          </span></span>
<span><span class="co">##  [17] S4Arrays_1.5.3              progress_1.2.3             </span></span>
<span><span class="co">##  [19] curl_5.2.1                  SparseArray_1.5.12         </span></span>
<span><span class="co">##  [21] Formula_1.2-5               sass_0.4.9                 </span></span>
<span><span class="co">##  [23] bslib_0.7.0                 htmlwidgets_1.6.4          </span></span>
<span><span class="co">##  [25] desc_1.4.3                  plyr_1.8.9                 </span></span>
<span><span class="co">##  [27] Gviz_1.49.0                 httr2_1.0.1                </span></span>
<span><span class="co">##  [29] cachem_1.1.0                GenomicAlignments_1.41.0   </span></span>
<span><span class="co">##  [31] lifecycle_1.0.4             pkgconfig_2.0.3            </span></span>
<span><span class="co">##  [33] Matrix_1.7-0                R6_2.5.1                   </span></span>
<span><span class="co">##  [35] fastmap_1.2.0               GenomeInfoDbData_1.2.12    </span></span>
<span><span class="co">##  [37] MatrixGenerics_1.17.0       digest_0.6.36              </span></span>
<span><span class="co">##  [39] colorspace_2.1-0            GGally_2.2.1               </span></span>
<span><span class="co">##  [41] AnnotationDbi_1.67.0        S4Vectors_0.43.1           </span></span>
<span><span class="co">##  [43] textshaping_0.4.0           Hmisc_5.1-3                </span></span>
<span><span class="co">##  [45] GenomicRanges_1.57.1        RSQLite_2.3.7              </span></span>
<span><span class="co">##  [47] filelock_1.0.3              fansi_1.0.6                </span></span>
<span><span class="co">##  [49] httr_1.4.7                  abind_1.4-5                </span></span>
<span><span class="co">##  [51] compiler_4.4.1              bit64_4.0.5                </span></span>
<span><span class="co">##  [53] pander_0.6.5                backports_1.5.0            </span></span>
<span><span class="co">##  [55] htmlTable_2.4.2             BiocParallel_1.39.0        </span></span>
<span><span class="co">##  [57] DBI_1.2.3                   ggstats_0.6.0              </span></span>
<span><span class="co">##  [59] biomaRt_2.61.2              rappdirs_0.3.3             </span></span>
<span><span class="co">##  [61] DelayedArray_0.31.5         rjson_0.2.21               </span></span>
<span><span class="co">##  [63] tools_4.4.1                 foreign_0.8-87             </span></span>
<span><span class="co">##  [65] nnet_7.3-19                 glue_1.7.0                 </span></span>
<span><span class="co">##  [67] restfulr_0.0.15             grid_4.4.1                 </span></span>
<span><span class="co">##  [69] checkmate_2.3.1             reshape2_1.4.4             </span></span>
<span><span class="co">##  [71] cluster_2.1.6               generics_0.1.3             </span></span>
<span><span class="co">##  [73] gtable_0.3.5                BSgenome_1.73.0            </span></span>
<span><span class="co">##  [75] ensembldb_2.29.0            tidyr_1.3.1                </span></span>
<span><span class="co">##  [77] data.table_1.15.4           hms_1.1.3                  </span></span>
<span><span class="co">##  [79] xml2_1.3.6                  utf8_1.2.4                 </span></span>
<span><span class="co">##  [81] XVector_0.45.0              BiocGenerics_0.51.0        </span></span>
<span><span class="co">##  [83] pillar_1.9.0                stringr_1.5.1              </span></span>
<span><span class="co">##  [85] limma_3.61.2                dplyr_1.1.4                </span></span>
<span><span class="co">##  [87] BiocFileCache_2.13.0        lattice_0.22-6             </span></span>
<span><span class="co">##  [89] deldir_2.0-4                rtracklayer_1.65.0         </span></span>
<span><span class="co">##  [91] bit_4.0.5                   biovizBase_1.53.0          </span></span>
<span><span class="co">##  [93] tidyselect_1.2.1            Biostrings_2.73.1          </span></span>
<span><span class="co">##  [95] knitr_1.47                  gridExtra_2.3              </span></span>
<span><span class="co">##  [97] bookdown_0.40               ProtGenerics_1.37.0        </span></span>
<span><span class="co">##  [99] IRanges_2.39.1              SummarizedExperiment_1.35.1</span></span>
<span><span class="co">## [101] stats4_4.4.1                xfun_0.45                  </span></span>
<span><span class="co">## [103] Biobase_2.65.0              statmod_1.5.0              </span></span>
<span><span class="co">## [105] matrixStats_1.3.0           stringi_1.8.4              </span></span>
<span><span class="co">## [107] UCSC.utils_1.1.0            lazyeval_0.2.2             </span></span>
<span><span class="co">## [109] yaml_2.3.8                  evaluate_0.24.0            </span></span>
<span><span class="co">## [111] codetools_0.2-20            interp_1.1-6               </span></span>
<span><span class="co">## [113] tibble_3.2.1                BiocManager_1.30.23        </span></span>
<span><span class="co">## [115] cli_3.6.3                   rpart_4.1.23               </span></span>
<span><span class="co">## [117] systemfonts_1.1.0           munsell_0.5.1              </span></span>
<span><span class="co">## [119] jquerylib_0.1.4             dichromat_2.0-0.1          </span></span>
<span><span class="co">## [121] Rcpp_1.0.12                 GenomeInfoDb_1.41.1        </span></span>
<span><span class="co">## [123] dbplyr_2.5.0                png_0.1-8                  </span></span>
<span><span class="co">## [125] XML_3.99-0.17               parallel_4.4.1             </span></span>
<span><span class="co">## [127] pkgdown_2.0.9.9000          ggplot2_3.5.1              </span></span>
<span><span class="co">## [129] blob_1.2.4                  prettyunits_1.2.0          </span></span>
<span><span class="co">## [131] jpeg_0.1-10                 latticeExtra_0.6-30        </span></span>
<span><span class="co">## [133] AnnotationFilter_1.29.0     bitops_1.0-7               </span></span>
<span><span class="co">## [135] VariantAnnotation_1.51.0    scales_1.3.0               </span></span>
<span><span class="co">## [137] purrr_1.0.2                 crayon_1.5.3               </span></span>
<span><span class="co">## [139] rlang_1.1.4                 KEGGREST_1.45.1</span></span></code></pre>
</div>
<div class="section level2">
<h2 class="unnumbered" id="References">References<a class="anchor" aria-label="anchor" href="#References"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" entry-spacing="0">
<div id="ref-RN1" class="csl-entry">
McLaren, W., B. Pritchard, D. Rios, Y. Chen, P. Flicek, and F.
Cunningham. 2010. <span>“Deriving the Consequences of Genomic Variants
with the Ensembl <span>API</span> and <span>SNP</span> Effect
Predictor.”</span> Journal Article. <em>Bioinformatics</em> 26 (16):
2069–70. <a href="https://doi.org/10.1093/bioinformatics/btq330" class="external-link">https://doi.org/10.1093/bioinformatics/btq330</a>.
</div>
</div>
</div>
<div class="footnotes footnotes-end-of-document">
<hr>
<ol>
<li id="fn1"><p>This statement below would be more evident if the
<code>summary</code> method was displaying the result of
<code>evalSeparately</code> in this vignette as it does it in an R
session.<a href="#fnref1" class="footnote-back">↩︎</a></p></li>
<li id="fn2"><p>Again, this statement would benefit from the result of
<code>evalSeparately</code> being displayed identically to an R
session.<a href="#fnref2" class="footnote-back">↩︎</a></p></li>
</ol>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Kevin Rue-Albrecht.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.9.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
